{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import time\n",
    "import os\n",
    "import gdreg\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pickle\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "\n",
    "# autoreload\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Trait info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DIC_NAME = {\n",
    "    'blood_PLATELET_COUNT': 'Platelet Count', \n",
    "    'blood_RBC_DISTRIB_WIDTH': 'Red Blood Cell Distribution Width',\n",
    "    'blood_RED_COUNT': 'Red Blood Cell Count',\n",
    "    'blood_WHITE_COUNT': 'White Blood Cell Count', \n",
    "    'bmd_HEEL_TSCOREz': 'Heel T Score', \n",
    "    'body_BALDING1': 'Balding Type I', \n",
    "    'body_BMIz': 'BMI',\n",
    "    'body_HEIGHTz': 'Height',\n",
    "    'body_WHRadjBMIz': 'Waist-hip Ratio',\n",
    "    'bp_DIASTOLICadjMEDz': 'Diastolic Blood Pressure',\n",
    "    'cov_EDU_YEARS': 'Years of Education',\n",
    "    'disease_ALLERGY_ECZEMA_DIAGNOSED': 'Eczema',\n",
    "    'lung_FEV1FVCzSMOKE': 'FEV1-FVC Ratio',\n",
    "    'lung_FVCzSMOKE': 'Forced Vital Capacity (FVC)',\n",
    "    'mental_NEUROTICISM': 'Neuroticism',\n",
    "    'other_MORNINGPERSON': 'Morning Person',\n",
    "    'pigment_SUNBURN': 'Sunburn Occasion',\n",
    "    'repro_MENARCHE_AGE': 'Age at Menarche',\n",
    "    'repro_MENOPAUSE_AGE': 'Age at Menopause',\n",
    "    'repro_NumberChildrenEverBorn_Pooled': 'Number Children (Pooled)',\n",
    "    'cancer_BREAST' : 'Breast Cancer',\n",
    "    'cancer_PROSTATE' : 'Prostate Cancer',\n",
    "    'disease_HYPOTHYROIDISM_SELF_REP' : 'Hypothyroidism',\n",
    "    'biochemistry_AlkalinePhosphatase' : 'Alkaline Phosphatase',\n",
    "    'biochemistry_AspartateAminotransferase' : 'Aspartate Aminotransferase',\n",
    "    'biochemistry_Cholesterol' : 'Cholesterol', \n",
    "    'biochemistry_Creatinine' : 'Creatinine',\n",
    "    'biochemistry_IGF1' : 'IGF1',\n",
    "    'biochemistry_Phosphate' : 'Phosphate',\n",
    "    'biochemistry_TotalBilirubin' : 'Total Bilirubin',\n",
    "    'biochemistry_TotalProtein' : 'Total Protein',\n",
    "    'biochemistry_VitaminD' : 'Vitamin D',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_info = pd.DataFrame(index=list(DIC_NAME))\n",
    "df_info['Trait Name'] = [DIC_NAME[x] for x in df_info.index]\n",
    "df_info['Trait_Identifier'] = df_info.index\n",
    "\n",
    "sumstats_file = '/n/groups/price/martin/data_GDREG/UKBimp_337K_MAF001/UKBB_trait/@.sumstats.gz'\n",
    "df_info['N.ukbimp'] = [pd.read_csv(sumstats_file.replace('@', x), nrows=10, sep='\\t')['N'].median()\n",
    "                for x in df_info.index]\n",
    "df_info['N.ukbimp'] = df_info['N.ukbimp'].astype(int)\n",
    "\n",
    "df_info.to_csv('/n/groups/price/martin/data_GDREG/results/tables/trait_info.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Single-SNP annotation info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# [Name, type, reference]\n",
    "DIC_NAME_AN = {\n",
    "    'BivFlnk' : ['Flanking bivalent TSS/enhancer', 'b', 'Roadmap Consortium 2015'],\n",
    "    'Conserved_LindbladToh' : ['Conserved (Lindblad-Toh)', 'b', 'Ward and Kellis 2012'],\n",
    "    'Conserved_Vertebrate_phastCons46way' : ['Conserved in vertebrates (phastCons)', 'b', 'Siepel et al. 2005'],\n",
    "    'Conserved_Mammal_phastCons46way' : ['Conserved in mammals (phastCons)', 'b', 'Siepel et al. 2005'],\n",
    "    'Conserved_Primate_phastCons46way' : ['Conserved in primates (phastCons)', 'b', 'Siepel et al. 2005'],\n",
    "    'Coding_UCSC' : ['Coding', 'b', 'UCSC'],\n",
    "    'UTR_3_UCSC' : ['3UTR', 'b', 'UCSC'],\n",
    "    'UTR_5_UCSC' : ['5UTR', 'b', 'UCSC'],\n",
    "    'Intron_UCSC' : ['Intron', 'b', 'UCSC'],\n",
    "    'Enhancer_Andersson' : ['FANTOM5 enhancer', 'b', 'Andersson et al. 2014'],\n",
    "    'SuperEnhancer_Hnisz' : ['Super enhancer (Hnisz)', 'b', 'Hnisz et al. 2013'],\n",
    "    'Enhancer_Hoffman' : ['Enhancer', 'b', 'Hoffman et al. 2013'],\n",
    "    'WeakEnhancer_Hoffman' : ['Weak enhancer', 'b', 'Hoffman et al. 2013'],\n",
    "    'Promoter_UCSC' : ['Promoter', 'b', 'UCSC'],\n",
    "    'PromoterFlanking_Hoffman' : ['Promoter flanking', 'b', 'Hoffman et al. 2013'],\n",
    "    'DHS_Trynka' : ['DHS', 'b', 'Trynka et al. 2013'],\n",
    "    'FetalDHS_Trynka' : ['Fetal DHS', 'b', 'Trynka et al. 2013'],\n",
    "    'H3K27ac_Hnisz' : ['H3K27ac (Hnisz)', 'b', 'Hnisz et al. 2013'],\n",
    "    'H3K27ac_PGC2' : ['H3K27ac (PGC2)', 'b', 'PGC 2014'],\n",
    "    'H3K4me1_Trynka' : ['H3K4me1', 'b', 'Trynka et al. 2013'],\n",
    "    'H3K4me3_Trynka' : ['H3K4me3', 'b', 'Trynka et al. 2013'],\n",
    "    'H3K9ac_Trynka' : ['H3K9ac', 'b', 'Trynka et al. 2013'],\n",
    "    'TFBS_ENCODE' : ['TFBS', 'b', 'ENCODE'],\n",
    "    'Transcribed_Hoffman' : ['Transcribed', 'b', 'Hoffman et al. 2013'],\n",
    "    'TSS_Hoffman' : ['TSS', 'b', 'Hoffman et al. 2013'],\n",
    "    'CTCF_Hoffman' : ['CTCF', 'b', 'Hoffman et al. 2013'],\n",
    "    'DGF_ENCODE' : ['DGF', 'b', 'ENCODE'],\n",
    "    'Repressed_Hoffman' : ['Repressed', 'b', 'Hoffman et al. 2013'],\n",
    "    'Vahedi_Tcell_SE' : ['Super enhancer (Vahedi)', 'b', 'Vahedi et al. 2015'],\n",
    "    'Vahedi_Tcell_TE' : ['Typical enhancer', 'b', 'Vahedi et al. 2015'],\n",
    "    'GerpRS_g4' : ['Conserved (GERP RS >= 4)', 'b', 'Davydov et al. 2010'], # from CADD\n",
    "    'GerpN' : ['Conserved (GERP NS)', 'c', 'Davydov et al. 2010'], # from CADD\n",
    "    'alleleage' : ['MAF-adjusted predicted allele age', 'c', 'Rasmussen et al. 2014'],\n",
    "    'LLD_AFR' : ['MAF-adjusted LLD-AFR', 'c', 'Gazal et al. 2017'],\n",
    "    'nucleotide_div' : ['Nucleotide diversity', 'c', ''],\n",
    "    'recomb_rate' : ['Recombination rate', 'c', 'HapMap2 map'],\n",
    "    'Backgrd_Selection_Stat' : ['McVicker B statistics', 'c', 'McVicker et al. 2009'],\n",
    "    'CpG' : ['CpG content', 'c', ''], # from CADD\n",
    "    # Non-synonymous removed\n",
    "    # Synonymous removed\n",
    "    'CADD_g20' : ['Deleterious (CADD >= 20)', 'b', 'Kircher et al. 2014'],\n",
    "    'snpeff_nonsynonymous_variant' : ['Non-synonymous', 'b', 'Cingolani et al. 2012'],\n",
    "    'snpeff_synonymous_variant' : ['Synonymous', 'b', 'Cingolani et al. 2012'],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_type(line):\n",
    "    if line == 'b':\n",
    "        return 'Binary'\n",
    "    if line == 'c':\n",
    "        return 'Continuous'\n",
    "\n",
    "df_info = pd.DataFrame(index=DIC_NAME_AN)\n",
    "df_info['Name'] = [DIC_NAME_AN[x][0] for x in df_info.index]\n",
    "df_info['Identifier'] = df_info.index\n",
    "df_info['Type'] = [get_type(DIC_NAME_AN[x][1]) for x in df_info.index]\n",
    "df_info['n_snp_common'] = 0\n",
    "df_info['n_snp_lf'] = 0\n",
    "df_info['Reference/dataset'] = [DIC_NAME_AN[x][2] for x in df_info.index]\n",
    "\n",
    "# n_snp\n",
    "temp_df = pd.read_csv(\n",
    "    '/n/groups/price/martin/data_GDREG/UKBimp_337K_MAF001/gdreg_res.baseline_100/blood_PLATELET_COUNT.tau.tsv',\n",
    "    sep='\\t', index_col=0\n",
    ")\n",
    "for row in df_info.index:\n",
    "    if df_info.loc[row, 'Type'] == 'Binary':\n",
    "        df_info.loc[row, 'n_snp_common'] = temp_df.loc['AN:%s_common'%row, 'n_snp']\n",
    "        df_info.loc[row, 'n_snp_lf'] = temp_df.loc['AN:%s_lf'%row, 'n_snp']\n",
    "df_info['n_snp_common'] = df_info['n_snp_common'].astype(int)\n",
    "df_info['n_snp_lf'] = df_info['n_snp_lf'].astype(int)\n",
    "\n",
    "# Version of the baseline model\n",
    "df_ldsc = pd.read_excel('/n/groups/price/martin/data_GDREG/results/Gazal_NG_2018_table.xlsx',\n",
    "                        sheet_name='S1', skiprows=1)\n",
    "df_ldsc.index = df_ldsc['Main annotation']\n",
    "df_ldsc.loc[df_ldsc['Version of the baseline model']=='New annotation', 'Version of the baseline model'] = \\\n",
    "    'Gazal et al. 2018' \n",
    "df_info['Version of the baseline model'] = [\n",
    "    df_ldsc.loc[x, 'Version of the baseline model'] \n",
    "    if (x in df_ldsc.index) & (x not in ['Non-synonymous', 'Synonymous']) \n",
    "    else 'New annotation' \n",
    "    for x in df_info['Name']\n",
    "]\n",
    "\n",
    "df_info.to_csv('/n/groups/price/martin/data_GDREG/results/tables/annot_info.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SNP-pair annotation info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "DIC_NAME_pAN = {\n",
    "    'proxy_0_100' : ['Proximal (0-0.1 kb)', 'All pairs of SNPs within 100 bp'],\n",
    "    'proxy_0_1000' : ['Proximal (0-1 kb)', 'All pairs of SNPs within 1 kb'],\n",
    "    'proxy_0_10000' : ['Proximal (0-10 kb)', 'All pairs of SNPs within 10 kb'],\n",
    "    'proxy_100_1000' : ['Proximal (0.1-1 kb)', 'All pairs of SNPs between 100 bp and 1 kb'],\n",
    "    'proxy_1000_10000' : ['Proximal (1-10 kb)', 'All pairs of SNPs between 1 kb and 10 kb'],\n",
    "    'ldp5_proxy_10000' : ['High-LD', 'All pairs of SNPs with LD >= 0.5 and dist < 10 kb'],\n",
    "    'ldp5_proxy_0_100' : ['High-LD (0-0.1 kb)', 'All pairs of SNPs with LD >= 0.5 and dist < 100 bp'],\n",
    "    'ldp5_proxy_100_1000' : ['High-LD (0.1-1 kb)', \n",
    "                             'All pairs of SNPs with LD >= 0.5 and dist between 100 bp and 1 kb'],\n",
    "    'ldp5_proxy_1000_10000' : ['High-LD (1-10 kb)', \n",
    "                                      'All pairs of SNPs with LD >= 0.5 and dist between 1 kb and 10 kb'],\n",
    "    'exon' : ['Exon', 'All pairs of SNPs in the same exon'],\n",
    "    'gene' : ['Gene', 'All pairs of SNPs in the same gene'],\n",
    "    'exonic_gene' : ['Exonic-gene', 'All pair of exonic SNPs in the same gene body'],\n",
    "    'protein_domain' : ['Protein-domain', 'All pair of SNPs in the same protein domain and same gene'],\n",
    "    'cS2G_all' : ['cS2G-all', 'All pairs of SNPs linked to the same gene by cS2G'],\n",
    "#     'cS2G_promoter' : ['cS2G-promoter', 'All pairs of promoter SNPs linked to the same gene by cS2G'],\n",
    "    'cS2G_promoter' : ['Promoter', 'All pairs of promoter SNPs linked to the same gene by cS2G'],\n",
    "    'cS2G_other' : [\n",
    "        'cS2G-other', \n",
    "        'All pairs of other SNPs linked to the same gene by cS2G,  excluding exonic or promoter SNPs'],\n",
    "    'Dixon_2015_H1_ESC' : ['Loop-ESC', 'Chromatin loops in H1 human embryonic stem (ES) cells'],\n",
    "    'Dixon_2015_H1_MES' : ['Loop-MES', 'Chromatin loops in H1 human mesendoderm (ME) cell'],\n",
    "    'Dixon_2015_H1_MSC' : ['Loop-MSC', 'Chromatin loops in H1 human mesenchymal stem (MS) cell'],\n",
    "    'Dixon_2015_H1_NPC' : ['Loop-NPC', 'Chromatin loops in H1 human neural progenitor (NP) cell'],\n",
    "    'Dixon_2015_H1_TRO' : ['Loop-TRO', 'Chromatin loops in H1 human trophoblast-like (TB) cell'],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_info = pd.DataFrame(index=DIC_NAME_pAN)\n",
    "df_info['Name'] = [DIC_NAME_pAN[x][0] for x in df_info.index]\n",
    "df_info['Identifier'] = df_info.index\n",
    "df_info['n_pair_common_common'] = 0\n",
    "df_info['n_pair_common_lf'] = 0\n",
    "df_info['n_pair_lf_lf'] = 0\n",
    "df_info['Description'] = [DIC_NAME_pAN[x][1] for x in df_info.index]\n",
    "\n",
    "# n_pair\n",
    "STATS_FILE = '/n/groups/price/martin/data_GDREG/UKBimp_337K_MAF001/pannot/stats/@.dist.tsv'\n",
    "for row in df_info.index:\n",
    "    for mbin in ['common_common', 'common_lf', 'lf_lf']:\n",
    "        if os.path.exists(STATS_FILE.replace('@', '%s_%s' % (row,mbin))):\n",
    "            temp_df = pd.read_csv(STATS_FILE.replace('@', '%s_%s' % (row,mbin)), sep='\\t')\n",
    "            df_info.loc[row, 'n_pair_%s' % mbin] = temp_df['count'].sum()\n",
    "            \n",
    "df_info.to_csv('/n/groups/price/martin/data_GDREG/results/tables/pannot_info.tsv', sep='\\t', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
