{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import time\n",
    "import os\n",
    "import gdreg\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pickle\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "\n",
    "# autoreload\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Trait info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DIC_NAME = {\n",
    "    'blood_PLATELET_COUNT': 'Platelet Count', \n",
    "    'blood_RBC_DISTRIB_WIDTH': 'Red Blood Cell Distribution Width',\n",
    "    'blood_RED_COUNT': 'Red Blood Cell Count',\n",
    "    'blood_WHITE_COUNT': 'White Blood Cell Count', \n",
    "    'bmd_HEEL_TSCOREz': 'Heel T Score', \n",
    "    'body_BALDING1': 'Balding Type I', \n",
    "    'body_BMIz': 'BMI',\n",
    "    'body_HEIGHTz': 'Height',\n",
    "    'body_WHRadjBMIz': 'Waist-hip Ratio',\n",
    "    'bp_DIASTOLICadjMEDz': 'Diastolic Blood Pressure',\n",
    "    'cov_EDU_YEARS': 'Years of Education',\n",
    "    'disease_ALLERGY_ECZEMA_DIAGNOSED': 'Eczema',\n",
    "    'lung_FEV1FVCzSMOKE': 'FEV1-FVC Ratio',\n",
    "    'lung_FVCzSMOKE': 'Forced Vital Capacity (FVC)',\n",
    "    'mental_NEUROTICISM': 'Neuroticism',\n",
    "    'other_MORNINGPERSON': 'Morning Person',\n",
    "    'pigment_SUNBURN': 'Sunburn Occasion',\n",
    "    'repro_MENARCHE_AGE': 'Age at Menarche',\n",
    "    'repro_MENOPAUSE_AGE': 'Age at Menopause',\n",
    "    'repro_NumberChildrenEverBorn_Pooled': 'Number Children (Pooled)',\n",
    "    'cancer_BREAST' : 'Breast Cancer',\n",
    "    'cancer_PROSTATE' : 'Prostate Cancer',\n",
    "    'disease_HYPOTHYROIDISM_SELF_REP' : 'Hypothyroidism',\n",
    "    'biochemistry_AlkalinePhosphatase' : 'Alkaline Phosphatase',\n",
    "    'biochemistry_AspartateAminotransferase' : 'Aspartate Aminotransferase',\n",
    "    'biochemistry_Cholesterol' : 'Cholesterol', \n",
    "    'biochemistry_Creatinine' : 'Creatinine',\n",
    "    'biochemistry_IGF1' : 'IGF1',\n",
    "    'biochemistry_Phosphate' : 'Phosphate',\n",
    "    'biochemistry_TotalBilirubin' : 'Total Bilirubin',\n",
    "    'biochemistry_TotalProtein' : 'Total Protein',\n",
    "    'biochemistry_VitaminD' : 'Vitamin D',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Avg. N = 299410\n"
     ]
    }
   ],
   "source": [
    "# ['Trait Name', 'Trait_Identifier', 'N.ukbimp']\n",
    "df_info = pd.DataFrame(index=list(DIC_NAME))\n",
    "df_info['Trait Name'] = [DIC_NAME[x] for x in df_info.index]\n",
    "df_info['Trait_Identifier'] = df_info.index\n",
    "\n",
    "trait_list = list(pd.read_csv(\n",
    "    '/n/groups/price/martin/data_GDREG/UKBimp_337K_MAF001/UKBB_trait/trait_list_all.txt', header=None,\n",
    ")[0])\n",
    "df_info = df_info.loc[trait_list].copy()\n",
    "\n",
    "sumstats_file = '/n/groups/price/martin/data_GDREG/UKBimp_337K_MAF001/UKBB_trait/@.sumstats.gz'\n",
    "df_info['N.ukbimp'] = [pd.read_csv(sumstats_file.replace('@', x), nrows=10, sep='\\t')['N'].median()\n",
    "                for x in df_info.index]\n",
    "df_info['N.ukbimp'] = df_info['N.ukbimp'].astype(int)\n",
    "\n",
    "df_info.to_csv('/n/groups/price/martin/data_GDREG/results/tables/trait_info.tsv', sep='\\t', index=False)\n",
    "print('Avg. N = %d' % df_info['N.ukbimp'].mean())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Single-SNP annotation info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# [Name, type, reference]\n",
    "DIC_NAME_AN = {\n",
    "    'BivFlnk' : ['Flanking bivalent TSS/enhancer', 'b', 'Roadmap Consortium 2015'],\n",
    "    'Conserved_LindbladToh' : ['Conserved (Lindblad-Toh)', 'b', 'Ward and Kellis 2012'],\n",
    "    'Conserved_Vertebrate_phastCons46way' : ['Conserved in vertebrates (phastCons)', 'b', 'Siepel et al. 2005'],\n",
    "    'Conserved_Mammal_phastCons46way' : ['Conserved in mammals (phastCons)', 'b', 'Siepel et al. 2005'],\n",
    "    'Conserved_Primate_phastCons46way' : ['Conserved in primates (phastCons)', 'b', 'Siepel et al. 2005'],\n",
    "    'Coding_UCSC' : ['Coding', 'b', 'UCSC'],\n",
    "    'UTR_3_UCSC' : ['3UTR', 'b', 'UCSC'],\n",
    "    'UTR_5_UCSC' : ['5UTR', 'b', 'UCSC'],\n",
    "    'Intron_UCSC' : ['Intron', 'b', 'UCSC'],\n",
    "    'Enhancer_Andersson' : ['FANTOM5 enhancer', 'b', 'Andersson et al. 2014'],\n",
    "    'SuperEnhancer_Hnisz' : ['Super enhancer (Hnisz)', 'b', 'Hnisz et al. 2013'],\n",
    "    'Enhancer_Hoffman' : ['Enhancer', 'b', 'Hoffman et al. 2013'],\n",
    "    'WeakEnhancer_Hoffman' : ['Weak enhancer', 'b', 'Hoffman et al. 2013'],\n",
    "    'Promoter_UCSC' : ['Promoter', 'b', 'UCSC'],\n",
    "    'PromoterFlanking_Hoffman' : ['Promoter flanking', 'b', 'Hoffman et al. 2013'],\n",
    "    'DHS_Trynka' : ['DHS', 'b', 'Trynka et al. 2013'],\n",
    "    'DHS_peaks_Trynka' : ['DHS peaks', 'b', 'Trynka et al. 2013'], # In baseline-LF but not in description\n",
    "    'FetalDHS_Trynka' : ['Fetal DHS', 'b', 'Trynka et al. 2013'],\n",
    "#     'H3K27ac_Hnisz' : ['H3K27ac (Hnisz)', 'b', 'Hnisz et al. 2013'], \n",
    "    'H3K27ac_Hnisz' : ['H3K27ac', 'b', 'Hnisz et al. 2013'], # Originally H3K27ac (Hnisz)\n",
    "    'H3K27ac_PGC2' : ['H3K27ac (PGC2)', 'b', 'PGC 2014'],\n",
    "    'H3K4me1_Trynka' : ['H3K4me1', 'b', 'Trynka et al. 2013'],\n",
    "    'H3K4me1_peaks_Trynka' : ['H3K4me1 peaks', 'b', 'Trynka et al. 2013'], # In baseline-LF but not in description\n",
    "    'H3K4me3_Trynka' : ['H3K4me3', 'b', 'Trynka et al. 2013'],\n",
    "    'H3K4me3_peaks_Trynka' : ['H3K4me3 peaks', 'b', 'Trynka et al. 2013'], # In baseline-LF but not in description\n",
    "    'H3K9ac_Trynka' : ['H3K9ac', 'b', 'Trynka et al. 2013'],\n",
    "    'H3K9ac_peaks_Trynka' : ['H3K9ac peaks', 'b', 'Trynka et al. 2013'], # In baseline-LF but not in description\n",
    "    'TFBS_ENCODE' : ['TFBS', 'b', 'ENCODE'],\n",
    "    'Transcribed_Hoffman' : ['Transcribed', 'b', 'Hoffman et al. 2013'],\n",
    "    'TSS_Hoffman' : ['TSS', 'b', 'Hoffman et al. 2013'],\n",
    "    'CTCF_Hoffman' : ['CTCF', 'b', 'Hoffman et al. 2013'],\n",
    "    'DGF_ENCODE' : ['DGF', 'b', 'ENCODE'],\n",
    "    'Repressed_Hoffman' : ['Repressed', 'b', 'Hoffman et al. 2013'],\n",
    "    'Vahedi_Tcell_SE' : ['Super enhancer (Vahedi)', 'b', 'Vahedi et al. 2015'],\n",
    "    'Vahedi_Tcell_TE' : ['Typical enhancer', 'b', 'Vahedi et al. 2015'],\n",
    "    'GerpRS_g4' : ['Conserved (GERP RS >= 4)', 'b', 'Davydov et al. 2010'], # from CADD\n",
    "    'GerpN' : ['Conserved (GERP NS)', 'c', 'Davydov et al. 2010'], # from CADD\n",
    "    'alleleage' : ['MAF-adjusted predicted allele age', 'c', 'Rasmussen et al. 2014'],\n",
    "    'LLD_AFR' : ['MAF-adjusted LLD-AFR', 'c', 'Gazal et al. 2017'],\n",
    "    'nucleotide_div' : ['Nucleotide diversity', 'c', ''],\n",
    "    'recomb_rate' : ['Recombination rate', 'c', 'HapMap2 map'],\n",
    "    'Backgrd_Selection_Stat' : ['McVicker B statistics', 'c', 'McVicker et al. 2009'],\n",
    "    'CpG' : ['CpG content', 'c', ''], # from CADD\n",
    "    # Non-synonymous removed\n",
    "    # Synonymous removed\n",
    "    'CADD_g20' : ['Deleterious (CADD >= 20)', 'b', 'Kircher et al. 2014'],\n",
    "    'snpeff_nonsynonymous_variant' : ['Non-synonymous', 'b', 'Cingolani et al. 2012'],\n",
    "    'snpeff_synonymous_variant' : ['Synonymous', 'b', 'Cingolani et al. 2012'],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_type(line):\n",
    "    if line == 'b':\n",
    "        return 'Binary'\n",
    "    if line == 'c':\n",
    "        return 'Continuous'\n",
    "\n",
    "df_info = pd.DataFrame(index=DIC_NAME_AN)\n",
    "df_info['Name'] = [DIC_NAME_AN[x][0] for x in df_info.index]\n",
    "df_info['Identifier'] = df_info.index\n",
    "df_info['Type'] = [get_type(DIC_NAME_AN[x][1]) for x in df_info.index]\n",
    "df_info['n_snp_common'] = 0\n",
    "df_info['n_snp_lf'] = 0\n",
    "df_info['Reference/dataset'] = [DIC_NAME_AN[x][2] for x in df_info.index]\n",
    "\n",
    "# n_snp\n",
    "temp_df = pd.read_csv(\n",
    "    '/n/groups/price/martin/data_GDREG/UKBimp_337K_MAF001/gdreg_res_021023.bsl1k_fct_nocl/'\n",
    "    'blood_PLATELET_COUNT.tau.tsv', sep='\\t', index_col=0\n",
    ")\n",
    "for row in df_info.index:\n",
    "    if df_info.loc[row, 'Type'] == 'Binary':\n",
    "        df_info.loc[row, 'n_snp_common'] = temp_df.loc['AN:%s_common'%row, 'n_snp']\n",
    "        df_info.loc[row, 'n_snp_lf'] = temp_df.loc['AN:%s_lf'%row, 'n_snp']\n",
    "df_info['n_snp_common'] = df_info['n_snp_common'].astype(int)\n",
    "df_info['n_snp_lf'] = df_info['n_snp_lf'].astype(int)\n",
    "\n",
    "# Version of the baseline model\n",
    "df_ldsc = pd.read_excel('/n/groups/price/martin/data_GDREG/results/Gazal_NG_2018_table.xlsx',\n",
    "                        sheet_name='S1', skiprows=1)\n",
    "df_ldsc.index = df_ldsc['Main annotation']\n",
    "df_ldsc.loc[df_ldsc['Version of the baseline model']=='New annotation', 'Version of the baseline model'] = \\\n",
    "    'Gazal et al. 2018' \n",
    "df_info['Version of the baseline model'] = [\n",
    "    df_ldsc.loc[x, 'Version of the baseline model'] \n",
    "    if (x in df_ldsc.index) & (x not in ['Non-synonymous', 'Synonymous']) \n",
    "    else 'New annotation' \n",
    "    for x in df_info['Name']\n",
    "]\n",
    "\n",
    "df_info.to_csv('/n/groups/price/martin/data_GDREG/results/tables/annot_info.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "### baseline-LF\n",
    "DATA_PATH =  '/n/groups/price/martin/data_GDREG/UKBimp_337K_MAF001'\n",
    "with open(DATA_PATH + '/gdreg_res.bsl_fct_100_nocl/blood_PLATELET_COUNT.pickle', 'br') as f:\n",
    "    temp_df = pickle.load(f)\n",
    "df_bsl = temp_df['summary']['tau'][['annot', 'type', 'n_snp']].copy()\n",
    "\n",
    "temp_list = ['AN:all'] + [x for x in df_bsl.index if ('mbin' in x) & ('common' in x)] + \\\n",
    "    [x for x in df_bsl.index if ('mbin' in x) & ('lf' in x)]\n",
    "temp_list += [x for x in df_bsl.index if (x not in temp_list) & (df_bsl.loc[x, 'type']=='non-binary')]\n",
    "temp_list += [x for x in df_bsl.index if (x not in temp_list)]\n",
    "df_bsl = df_bsl.loc[temp_list]\n",
    "\n",
    "df_bsl.loc[df_bsl['type']=='non-binary', 'n_snp'] = -1\n",
    "df_bsl['n_snp'] = df_bsl['n_snp'].astype(int)\n",
    "df_bsl.to_csv('/n/groups/price/martin/data_GDREG/results/tables/baseline_lf_165.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SNP-pair annotation info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n_annot=45\n"
     ]
    }
   ],
   "source": [
    "# Functional annotations\n",
    "DATA_PATH =  '/n/groups/price/martin/data_GDREG/UKBimp_337K_MAF001'\n",
    "dic_res = {}\n",
    "with open(DATA_PATH + '/gdreg_res.bsl_fct_100_nocl/blood_PLATELET_COUNT.pickle', 'br') as f:\n",
    "    dic_res['fct100'] = pickle.load(f)\n",
    "with open(DATA_PATH + '/gdreg_res.bsl_fct_1k_nocl/blood_PLATELET_COUNT.pickle', 'br') as f:\n",
    "    dic_res['fct1k'] = pickle.load(f)\n",
    "DF_ANNOT = pd.read_csv('/n/groups/price/martin/data_GDREG/results/tables/annot_info.tsv', sep='\\t')\n",
    "DF_ANNOT.index = DF_ANNOT['Identifier']\n",
    "print(\"n_annot=%d\" % DF_ANNOT.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "DIC_NAME_pAN = {\n",
    "    'proxy_0_100' : ['Proximal 0-100bp', '0-100bp', 'proximity'],\n",
    "    'proxy_0_1000' : ['Proximal 0-1kb', '0-1kb', 'proximity'],\n",
    "    'proxy_100_1000' : ['Proximal 100bp-1kb', '100bp-1kb', 'proximity'],\n",
    "    'exon' : ['Same-exon', 'Same exon', 'gene'],\n",
    "    'gene' : ['Same-gene', 'Same gene', 'gene'],\n",
    "    'exonic_gene' : ['Exonic', 'Exonic region of same gene', 'gene'],\n",
    "    'protein_domain' : ['Same-protein-domain', 'Same protein domain of same gene', 'gene'],\n",
    "    'cS2G_promoter' : ['Promoter', 'Promoter of same gene', 'gene'],\n",
    "}\n",
    "\n",
    "# Functional annotation\n",
    "for pAN in dic_res['fct100']['summary']['rho'].index:\n",
    "    pAN = pAN.replace('pAN:','').replace('_common_common','').replace('_lf_lf','')\n",
    "    if pAN not in DIC_NAME_pAN:\n",
    "        fct = pAN.replace('_0_100','')\n",
    "        DIC_NAME_pAN[pAN] = [\n",
    "            '%s-100'%DF_ANNOT.loc[fct, 'Name'], \n",
    "            '%s (0-100bp)' % DF_ANNOT.loc[fct, 'Name'], 'fct-100',\n",
    "        ]\n",
    "for pAN in dic_res['fct1k']['summary']['rho'].index:\n",
    "    pAN = pAN.replace('pAN:','').replace('_common_common','').replace('_lf_lf','')\n",
    "    if pAN not in DIC_NAME_pAN:\n",
    "        fct = pAN.replace('_0_1000','')\n",
    "        if fct not in DF_ANNOT.index:\n",
    "            print('Missing', fct)\n",
    "            continue\n",
    "        DIC_NAME_pAN[pAN] = [\n",
    "            '%s-1k'%DF_ANNOT.loc[fct, 'Name'], \n",
    "            '%s (0-1kb)' % DF_ANNOT.loc[fct, 'Name'], 'fct-1k',\n",
    "        ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_info = pd.DataFrame(index=DIC_NAME_pAN)\n",
    "df_info['Name'] = [DIC_NAME_pAN[x][0] for x in df_info.index]\n",
    "df_info['Identifier'] = df_info.index\n",
    "df_info['Type'] = [DIC_NAME_pAN[x][2] for x in df_info.index]\n",
    "df_info['n_pair_common_common'] = 0\n",
    "df_info['n_pair_common_lf'] = 0\n",
    "df_info['n_pair_lf_lf'] = 0\n",
    "df_info['avg_dist_common_common'] = 0\n",
    "df_info['avg_dist_common_lf'] = 0\n",
    "df_info['avg_dist_lf_lf'] = 0\n",
    "df_info['avgr_common_common'] = 0\n",
    "df_info['avgr_common_lf'] = 0\n",
    "df_info['avgr_lf_lf'] = 0\n",
    "df_info['Description'] = [DIC_NAME_pAN[x][1] for x in df_info.index]\n",
    "\n",
    "# n_pair & median distance\n",
    "STATS_FILE = '/n/groups/price/martin/data_GDREG/UKBimp_337K_MAF001/pannot/stats/@.dist.tsv'\n",
    "for row in df_info.index:\n",
    "    for mbin in ['common_common', 'common_lf', 'lf_lf']:\n",
    "        if os.path.exists(STATS_FILE.replace('@', '%s_%s' % (row, mbin))):\n",
    "            temp_df = pd.read_csv(STATS_FILE.replace('@', '%s_%s' % (row,mbin)), sep='\\t')\n",
    "            temp_df['mid'] = 0.5 * (temp_df['left'] + temp_df['right'])\n",
    "            temp_df['ct_prop'] = temp_df['count'] / temp_df['count'].sum()\n",
    "            df_info.loc[row, 'n_pair_%s' % mbin] = temp_df['count'].sum()\n",
    "            df_info.loc[row, 'avg_dist_%s' % mbin] = (temp_df['mid'] * temp_df['ct_prop']).sum()\n",
    "            \n",
    "# avgr \n",
    "AVGR_FILE = '/n/groups/price/martin/data_GDREG/UKBimp_337K_MAF001/pannot/main.avgr'\n",
    "temp_df = pd.read_csv(AVGR_FILE, sep='\\t', header=None)\n",
    "temp_df.index = temp_df[0]\n",
    "for row in df_info.index:\n",
    "    for mbin in ['common_common', 'common_lf', 'lf_lf']:\n",
    "        df_info.loc[row, 'avgr_%s' % mbin] = temp_df.loc['pAN:%s_%s' % (row, mbin), 1]\n",
    "            \n",
    "df_info.to_csv('/n/groups/price/martin/data_GDREG/results/tables/pannot_info.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n_pannot=32\n",
      "\\begin{tabular}{llllll}\n",
      "\\toprule\n",
      "                      Name &                       Description &  Size (C) & Size (LF) & Avg. Dist (C) & Avg. Dist (LF) \\\\\n",
      "\\midrule\n",
      "            Proximal 0-1kb &                             0-1kb &    23.62M &     8.46M &           490 &            487 \\\\\n",
      "                 Same-exon &                         Same exon &     0.54M &     0.26M &          3598 &           3744 \\\\\n",
      "                 Same-gene &                         Same gene &  1367.49M &   521.53M &        392654 &         385553 \\\\\n",
      "                    Exonic &        Exonic region of same gene &     1.17M &     0.61M &         52431 &          57013 \\\\\n",
      "       Same-protein-domain &  Same protein domain of same gene &     0.09M &     0.09M &         41381 &          54120 \\\\\n",
      "                  Promoter &             Promoter of same gene &     0.88M &     0.37M &         44571 &          51104 \\\\\n",
      "                    DGF-1k &                       DGF (0-1kb) &     0.85M &     0.34M &           387 &            386 \\\\\n",
      "                    DHS-1k &                       DHS (0-1kb) &     1.38M &     0.53M &           383 &            384 \\\\\n",
      "              DHS peaks-1k &                 DHS peaks (0-1kb) &     0.65M &     0.26M &           366 &            368 \\\\\n",
      "               Enhancer-1k &                  Enhancer (0-1kb) &     0.52M &     0.20M &           419 &            415 \\\\\n",
      "              Fetal DHS-1k &                 Fetal DHS (0-1kb) &     0.60M &     0.24M &           348 &            351 \\\\\n",
      "                H3K27ac-1k &                   H3K27ac (0-1kb) &     8.53M &     3.11M &           484 &            481 \\\\\n",
      "         H3K27ac (PGC2)-1k &            H3K27ac (PGC2) (0-1kb) &     5.29M &     1.96M &           470 &            467 \\\\\n",
      "                H3K4me1-1k &                   H3K4me1 (0-1kb) &     7.47M &     2.79M &           467 &            465 \\\\\n",
      "          H3K4me1 peaks-1k &             H3K4me1 peaks (0-1kb) &     1.49M &     0.60M &           432 &            431 \\\\\n",
      "                H3K4me3-1k &                   H3K4me3 (0-1kb) &     1.92M &     0.77M &           437 &            436 \\\\\n",
      "                 H3K9ac-1k &                    H3K9ac (0-1kb) &     1.76M &     0.72M &           441 &            442 \\\\\n",
      "                 Intron-1k &                    Intron (0-1kb) &     8.29M &     3.18M &           487 &            485 \\\\\n",
      "               Promoter-1k &                  Promoter (0-1kb) &     0.95M &     0.37M &           470 &            467 \\\\\n",
      "              Repressed-1k &                 Repressed (0-1kb) &     8.20M &     2.76M &           466 &            463 \\\\\n",
      " Super enhancer (Hnisz)-1k &    Super enhancer (Hnisz) (0-1kb) &     3.89M &     1.40M &           488 &            485 \\\\\n",
      "                   TFBS-1k &                      TFBS (0-1kb) &     1.40M &     0.51M &           390 &            384 \\\\\n",
      "            Transcribed-1k &               Transcribed (0-1kb) &     4.85M &     1.97M &           458 &            459 \\\\\n",
      "       Typical enhancer-1k &          Typical enhancer (0-1kb) &     0.48M &     0.17M &           477 &            472 \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\n",
      "100bp pannots\n",
      "H3K27ac-100, H3K27ac (PGC2)-100, H3K4me1-100, Intron-100, Repressed-100, Transcribed-100\n",
      "\n",
      "1kb pannots\n",
      "DGF-1k, DHS-1k, DHS peaks-1k, Enhancer-1k, Fetal DHS-1k, H3K27ac-1k, H3K27ac (PGC2)-1k, H3K4me1-1k, H3K4me1 peaks-1k, H3K4me3-1k, H3K9ac-1k, Intron-1k, Promoter-1k, Repressed-1k, Super enhancer (Hnisz)-1k, TFBS-1k, Transcribed-1k, Typical enhancer-1k\n"
     ]
    }
   ],
   "source": [
    "# Table 1 pannots in baseline-SP\n",
    "DF_PANNOT = pd.read_csv('/n/groups/price/martin/data_GDREG/results/tables/pannot_info.tsv', sep='\\t')\n",
    "# DF_PANNOT.index = DF_PANNOT['Name']\n",
    "print(\"n_pannot=%d\" % DF_PANNOT.shape[0])\n",
    "DF_PANNOT = DF_PANNOT[[\n",
    "    'Name', 'Description', 'n_pair_common_common', 'n_pair_lf_lf',\n",
    "    'avg_dist_common_common', 'avg_dist_lf_lf',\n",
    "]]\n",
    "for col in ['n_pair_common_common', 'n_pair_lf_lf']:\n",
    "    DF_PANNOT[col] = ['%0.2fM' % (x/1e6) for x in DF_PANNOT[col]]\n",
    "for col in ['avg_dist_common_common', 'avg_dist_lf_lf']:\n",
    "    DF_PANNOT[col] = ['%d' % x for x in DF_PANNOT[col]]\n",
    "DF_PANNOT.columns = ['Name', 'Description', 'Size (C)', 'Size (LF)', 'Avg. Dist (C)', 'Avg. Dist (LF)']\n",
    "\n",
    "keep_list = ['Proximal 0-100bp', 'Proximal 0-1kb', 'Same-exon', 'Same-gene', 'Exonic', 'Same-protein-domain', \n",
    "             'Promoter', 'H3K27ac-100', 'H3K27ac-1k']\n",
    "keep_list = ['Proximal 0-1kb', 'Same-exon', 'Same-gene', 'Exonic', 'Same-protein-domain', 'Promoter']\n",
    "keep_list += [x for x in DF_PANNOT['Name'] if x.endswith('1k')]\n",
    "print(DF_PANNOT.loc[DF_PANNOT['Name'].isin(keep_list)].to_latex(index=False))\n",
    "\n",
    "print('100bp pannots')\n",
    "print((', '.join(DF_PANNOT.loc[[x.endswith('100') for x in DF_PANNOT['Name']], 'Name'])))\n",
    "print('')\n",
    "print('1kb pannots')\n",
    "print((', '.join(DF_PANNOT.loc[[x.endswith('1k') for x in DF_PANNOT['Name']], 'Name'])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'DGF-1k, DHS-1k, DHS peaks-1k, Enhancer-1k, Fetal DHS-1k, H3K27ac-1k, H3K27ac (PGC2)-1k, H3K4me1-1k, H3K4me1 peaks-1k, H3K4me3-1k, H3K9ac-1k, Intron-1k, Promoter-1k, Repressed-1k, Super enhancer (Hnisz)-1k, TFBS-1k, Transcribed-1k, Typical enhancer-1k'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "', '.join()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n_pannot=32\n",
      "\\begin{tabular}{llllll}\n",
      "\\toprule\n",
      "                Name &                        Description &  Size (C) & Size (LF) & Avg. Dist (C) & Avg. Dist (LF) \\\\\n",
      "\\midrule\n",
      "    Proximal 0-100bp &                            <100 bp &     2.62M &     0.98M &            47 &             46 \\\\\n",
      "      Proximal 0-1kb &                              <1 kb &    23.62M &     8.46M &           490 &            487 \\\\\n",
      "           Same-exon &                          Same exon &     0.54M &     0.26M &          3598 &           3744 \\\\\n",
      "           Same-gene &                          Same gene &  1367.49M &   521.53M &        392654 &         385553 \\\\\n",
      "              Exonic &         Exonic region of same gene &     1.17M &     0.61M &         52431 &          57013 \\\\\n",
      " Same-protein-domain &  Same protein domain and same gene &     0.09M &     0.09M &         41381 &          54120 \\\\\n",
      "            Promoter &              Promoter of same gene &     0.88M &     0.37M &         44571 &          51104 \\\\\n",
      "         H3K27ac-100 &                   H3K27ac (<100bp) &     0.99M &     0.38M &            47 &             45 \\\\\n",
      "          H3K27ac-1k &                     H3K27ac (<1kb) &     8.53M &     3.11M &           484 &            481 \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\n",
      "100bp pannots\n",
      "H3K27ac-100, H3K27ac (PGC2)-100, H3K4me1-100, Intron-100, Repressed-100, Transcribed-100\n",
      "\n",
      "1kb pannots\n",
      "DGF-1k, DHS-1k, DHS peaks-1k, Enhancer-1k, Fetal DHS-1k, H3K27ac-1k, H3K27ac (PGC2)-1k, H3K4me1-1k, H3K4me1 peaks-1k, H3K4me3-1k, H3K9ac-1k, Intron-1k, Promoter-1k, Repressed-1k, Super enhancer (Hnisz)-1k, TFBS-1k, Transcribed-1k, Typical enhancer-1k\n"
     ]
    }
   ],
   "source": [
    "# Table 1 pannots\n",
    "DF_PANNOT = pd.read_csv('/n/groups/price/martin/data_GDREG/results/tables/pannot_info.tsv', sep='\\t')\n",
    "# DF_PANNOT.index = DF_PANNOT['Name']\n",
    "print(\"n_pannot=%d\" % DF_PANNOT.shape[0])\n",
    "DF_PANNOT = DF_PANNOT[[\n",
    "    'Name', 'Description', 'n_pair_common_common', 'n_pair_lf_lf',\n",
    "    'avg_dist_common_common', 'avg_dist_lf_lf',\n",
    "]]\n",
    "for col in ['n_pair_common_common', 'n_pair_lf_lf']:\n",
    "    DF_PANNOT[col] = ['%0.2fM' % (x/1e6) for x in DF_PANNOT[col]]\n",
    "for col in ['avg_dist_common_common', 'avg_dist_lf_lf']:\n",
    "    DF_PANNOT[col] = ['%d' % x for x in DF_PANNOT[col]]\n",
    "DF_PANNOT.columns = ['Name', 'Description', 'Size (C)', 'Size (LF)', 'Avg. Dist (C)', 'Avg. Dist (LF)']\n",
    "\n",
    "keep_list = ['Proximal 0-100bp', 'Proximal 0-1kb', 'Same-exon', 'Same-gene', 'Exonic', 'Same-protein-domain', \n",
    "             'Promoter', 'H3K27ac-100', 'H3K27ac-1k']\n",
    "print(DF_PANNOT.loc[DF_PANNOT['Name'].isin(keep_list)].to_latex(index=False))\n",
    "\n",
    "print('100bp pannots')\n",
    "print((', '.join(DF_PANNOT.loc[[x.endswith('100') for x in DF_PANNOT['Name']], 'Name'])))\n",
    "print('')\n",
    "print('1kb pannots')\n",
    "print((', '.join(DF_PANNOT.loc[[x.endswith('1k') for x in DF_PANNOT['Name']], 'Name'])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Old code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#########\n",
      "df_snp_chr (199680, 4)\n",
      "   CHR          SNP        BP       MAF\n",
      "0   22  rs587769434  16050527  0.001391\n",
      "1   22   rs62224609  16051249  0.100340\n",
      "2   22   rs62224614  16053862  0.101977\n",
      "#########\n",
      "df_cs2g (113016, 9)\n",
      "Overlap with df_snp_chr: 68926\n",
      "                  SNP   GENE  cS2G     INFO                  ID  Link  Exon  \\\n",
      "0         rs556077431  POTEH   1.0  |Exon=1     22:16258211_A_G  True  True   \n",
      "1  22:16258290_CATT_C  POTEH   1.0  |Exon=1  22:16258290_CATT_C  True  True   \n",
      "2         rs200891952  POTEH   1.0  |Exon=1     22:16258294_T_G  True  True   \n",
      "\n",
      "   Promoter  Other  \n",
      "0     False  False  \n",
      "1     False  False  \n",
      "2     False  False  \n",
      "Link        87789\n",
      "Exon        13072\n",
      "Promoter    11038\n",
      "Other       63679\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "PGEN_FILE = '/n/scratch3/users/j/jz286/imp_geno/ukb_imp_chr22_v3'\n",
    "PANNOT = 'cS2G_promoter'\n",
    "\n",
    "# df_snp_chr\n",
    "df_snp_chr = gdreg.util.read_pgen(PGEN_FILE)[\"pvar\"]\n",
    "df_snp_chr = df_snp_chr[[\"CHR\", \"SNP\", \"BP\"]].copy()\n",
    "CHR = df_snp_chr[\"CHR\"][0]\n",
    "df_snp_chr['MAF'] = gdreg.util.read_pgen(PGEN_FILE)[\"afreq\"][\"MAF\"].astype(np.float32)\n",
    "dic_bp = {x:y for x,y in zip(df_snp_chr[\"SNP\"], df_snp_chr[\"BP\"])}\n",
    "print('#########')\n",
    "print('df_snp_chr', df_snp_chr.shape)\n",
    "print(df_snp_chr.iloc[:3]) \n",
    "\n",
    "# df_cs2g\n",
    "df_cs2g = pd.read_csv(\n",
    "    '/n/groups/price/martin/data_GDREG/gene_annotation/cS2G/cS2G_UKBB/cS2G.%s.SGscore.gz' % CHR, sep='\\t',\n",
    ")\n",
    "df_snpmap = pd.read_csv(\n",
    "    '/n/groups/price/martin/data_GDREG/gene_annotation/cS2G/00_bim/UKBB.%s.info' % CHR, sep=' ',\n",
    ")\n",
    "\n",
    "temp_dic = {x:y for x,y in zip(df_snpmap['ID'], df_snpmap['RS'])}\n",
    "df_cs2g['ID'] = df_cs2g['SNP']\n",
    "df_cs2g['SNP'] = [temp_dic[x] for x in df_cs2g['ID']]\n",
    "\n",
    "df_cs2g['Link'] = df_cs2g['cS2G'] > 0.5\n",
    "df_cs2g['Exon'] = np.array(['Exon' in x for x in df_cs2g['INFO']]) & df_cs2g['Link']\n",
    "df_cs2g['Promoter'] = np.array(['Promoter' in x for x in df_cs2g['INFO']]) & df_cs2g['Link']\n",
    "df_cs2g['Other'] = (~df_cs2g['Exon']) & (~df_cs2g['Promoter']) & df_cs2g['Link']\n",
    "print('#########')\n",
    "print('df_cs2g', df_cs2g.shape)\n",
    "print('Overlap with df_snp_chr:', len(set(df_snp_chr['SNP']) & set(df_cs2g['SNP'])))\n",
    "print(df_cs2g.iloc[:3]) \n",
    "print(df_cs2g[['Link', 'Exon', 'Promoter', 'Other']].sum(axis=0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cS2G SNP pairs    \n",
    "if PANNOT == 'cS2G_all':\n",
    "    ind_select = df_cs2g['Link']\n",
    "\n",
    "if PANNOT == 'cS2G_promoter':\n",
    "    ind_select = df_cs2g['Link'] & df_cs2g['Promoter']\n",
    "\n",
    "if PANNOT == 'cS2G_other':\n",
    "    ind_select = df_cs2g['Link'] & df_cs2g['Other']\n",
    "\n",
    "temp_df = df_cs2g.loc[ind_select].copy()\n",
    "temp_df = temp_df.loc[temp_df['SNP'].isin(df_snp_chr['SNP'])]\n",
    "# temp_df = temp_df.groupby('GENE').agg({'SNP':list})  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp_df.index = temp_df['SNP']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_snp = df_snp_chr.copy()\n",
    "df_snp.index = df_snp['SNP']\n",
    "df_snp = df_snp.join(temp_df[['GENE', 'cS2G', 'INFO', 'ID', 'Link', 'Exon', 'Promoter', 'Other']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_snp = df_snp.loc[~df_snp['Promoter'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>SNP</th>\n",
       "      <th>BP</th>\n",
       "      <th>MAF</th>\n",
       "      <th>GENE</th>\n",
       "      <th>cS2G</th>\n",
       "      <th>INFO</th>\n",
       "      <th>ID</th>\n",
       "      <th>Link</th>\n",
       "      <th>Exon</th>\n",
       "      <th>Promoter</th>\n",
       "      <th>Other</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22:17101208_TG_T</th>\n",
       "      <td>22</td>\n",
       "      <td>22:17101208_TG_T</td>\n",
       "      <td>17101208</td>\n",
       "      <td>0.269351</td>\n",
       "      <td>TPTEP1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>|Promoter=1</td>\n",
       "      <td>22:17101208_TG_T</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22:17301606_AC_A</th>\n",
       "      <td>22</td>\n",
       "      <td>22:17301606_AC_A</td>\n",
       "      <td>17301606</td>\n",
       "      <td>0.129504</td>\n",
       "      <td>XKR3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>|Promoter=1</td>\n",
       "      <td>22:17301606_AC_A</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22:17302415_AT_A</th>\n",
       "      <td>22</td>\n",
       "      <td>22:17302415_AT_A</td>\n",
       "      <td>17302415</td>\n",
       "      <td>0.119405</td>\n",
       "      <td>XKR3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>|Promoter=1|ABC=1</td>\n",
       "      <td>22:17302415_AT_A</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22:17488801_TC_T</th>\n",
       "      <td>22</td>\n",
       "      <td>22:17488801_TC_T</td>\n",
       "      <td>17488801</td>\n",
       "      <td>0.001602</td>\n",
       "      <td>GAB4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>|Promoter=1|Cicero=1</td>\n",
       "      <td>22:17488801_TC_T</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22:17679877_GAA_G</th>\n",
       "      <td>22</td>\n",
       "      <td>22:17679877_GAA_G</td>\n",
       "      <td>17679877</td>\n",
       "      <td>0.042018</td>\n",
       "      <td>CECR1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>|Promoter=1|ABC=1</td>\n",
       "      <td>22:17679877_GAA_G</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs986643</th>\n",
       "      <td>22</td>\n",
       "      <td>rs986643</td>\n",
       "      <td>22397907</td>\n",
       "      <td>0.393497</td>\n",
       "      <td>PRAMENP</td>\n",
       "      <td>1.0</td>\n",
       "      <td>|Promoter=1|GTeX_Finemapped=1|eQTLGen_Finemapp...</td>\n",
       "      <td>22:22397907_T_C</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs986644</th>\n",
       "      <td>22</td>\n",
       "      <td>rs986644</td>\n",
       "      <td>22397943</td>\n",
       "      <td>0.042051</td>\n",
       "      <td>PRAMENP</td>\n",
       "      <td>1.0</td>\n",
       "      <td>|Promoter=1</td>\n",
       "      <td>22:22397943_C_G</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs986645</th>\n",
       "      <td>22</td>\n",
       "      <td>rs986645</td>\n",
       "      <td>22398127</td>\n",
       "      <td>0.393269</td>\n",
       "      <td>PRAMENP</td>\n",
       "      <td>1.0</td>\n",
       "      <td>|Promoter=1|GTeX_Finemapped=1|eQTLGen_Finemapp...</td>\n",
       "      <td>22:22398127_C_T</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs986646</th>\n",
       "      <td>22</td>\n",
       "      <td>rs986646</td>\n",
       "      <td>22398220</td>\n",
       "      <td>0.042089</td>\n",
       "      <td>PRAMENP</td>\n",
       "      <td>1.0</td>\n",
       "      <td>|Promoter=1</td>\n",
       "      <td>22:22398220_T_C</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs9941915</th>\n",
       "      <td>22</td>\n",
       "      <td>rs9941915</td>\n",
       "      <td>44420859</td>\n",
       "      <td>0.284862</td>\n",
       "      <td>PARVB</td>\n",
       "      <td>1.0</td>\n",
       "      <td>|Promoter=1|ABC=1</td>\n",
       "      <td>22:44420859_A_T</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8257 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   CHR                SNP        BP       MAF     GENE  cS2G  \\\n",
       "SNP                                                                            \n",
       "22:17101208_TG_T    22   22:17101208_TG_T  17101208  0.269351   TPTEP1   1.0   \n",
       "22:17301606_AC_A    22   22:17301606_AC_A  17301606  0.129504     XKR3   1.0   \n",
       "22:17302415_AT_A    22   22:17302415_AT_A  17302415  0.119405     XKR3   1.0   \n",
       "22:17488801_TC_T    22   22:17488801_TC_T  17488801  0.001602     GAB4   1.0   \n",
       "22:17679877_GAA_G   22  22:17679877_GAA_G  17679877  0.042018    CECR1   1.0   \n",
       "...                ...                ...       ...       ...      ...   ...   \n",
       "rs986643            22           rs986643  22397907  0.393497  PRAMENP   1.0   \n",
       "rs986644            22           rs986644  22397943  0.042051  PRAMENP   1.0   \n",
       "rs986645            22           rs986645  22398127  0.393269  PRAMENP   1.0   \n",
       "rs986646            22           rs986646  22398220  0.042089  PRAMENP   1.0   \n",
       "rs9941915           22          rs9941915  44420859  0.284862    PARVB   1.0   \n",
       "\n",
       "                                                                INFO  \\\n",
       "SNP                                                                    \n",
       "22:17101208_TG_T                                         |Promoter=1   \n",
       "22:17301606_AC_A                                         |Promoter=1   \n",
       "22:17302415_AT_A                                   |Promoter=1|ABC=1   \n",
       "22:17488801_TC_T                                |Promoter=1|Cicero=1   \n",
       "22:17679877_GAA_G                                  |Promoter=1|ABC=1   \n",
       "...                                                              ...   \n",
       "rs986643           |Promoter=1|GTeX_Finemapped=1|eQTLGen_Finemapp...   \n",
       "rs986644                                                 |Promoter=1   \n",
       "rs986645           |Promoter=1|GTeX_Finemapped=1|eQTLGen_Finemapp...   \n",
       "rs986646                                                 |Promoter=1   \n",
       "rs9941915                                          |Promoter=1|ABC=1   \n",
       "\n",
       "                                  ID  Link   Exon Promoter  Other  \n",
       "SNP                                                                \n",
       "22:17101208_TG_T    22:17101208_TG_T  True  False     True  False  \n",
       "22:17301606_AC_A    22:17301606_AC_A  True  False     True  False  \n",
       "22:17302415_AT_A    22:17302415_AT_A  True  False     True  False  \n",
       "22:17488801_TC_T    22:17488801_TC_T  True  False     True  False  \n",
       "22:17679877_GAA_G  22:17679877_GAA_G  True  False     True  False  \n",
       "...                              ...   ...    ...      ...    ...  \n",
       "rs986643             22:22397907_T_C  True  False     True  False  \n",
       "rs986644             22:22397943_C_G  True  False     True  False  \n",
       "rs986645             22:22398127_C_T  True  False     True  False  \n",
       "rs986646             22:22398220_T_C  True  False     True  False  \n",
       "rs9941915            22:44420859_A_T  True  False     True  False  \n",
       "\n",
       "[8257 rows x 12 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_snp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_exon = pd.read_csv(\n",
    "    \"/n/groups/price/martin/data_GDREG/gene_annotation/ENSE_exon_annot_v41.txt\", sep=\"\\t\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAsMAAAE8CAYAAAAloLwTAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAAS+UlEQVR4nO3de4yld33f8c83XkgICDdlFxR5vaxFTBsCwSlTHJU2obk6JrJJm0Q2uUBrcKMaclEvASkJFlGlBKnFQuGiLUEOiWzHJRFxCAlECYqbJnYZK0BsI8yWm9eleGPAyA1NsvDtH3PWexhmdmZ3zmVmf6+XNPKc5/md5/yOf/vsvPeZM3OquwMAACP6qmVPAAAAlkUMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwrH3LeuD9+/f34cOHl/XwAAAM4q677vqr7j6w0b6lxfDhw4ezurq6rIcHAGAQVfWJzfZ5mQQAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMPaMoar6q1V9WBV3b3FuH9cVSeq6gdnNz0AAJif7VwZvjHJZacbUFXnJfnlJO+ZwZwAAGAhtozh7r49yWe2GPaKJL+V5MFZTAoAABZhx68ZrqoLkvxAkjftfDoAALA4s/gBuhuS/Gx3f2mrgVV1bVWtVtXq8ePHZ/DQAABw9vbN4BgrSW6pqiTZn+TyqjrR3e9YP7C7jyQ5kiQrKys9g8cGAICztuMY7u6LTn5eVTcmeedGIQwAALvNljFcVTcneX6S/VV1LMmrkzwmSbr7zXOdHQAAzNGWMdzdV2/3YN39kh3NBgAAFsg70AEAMCwxDADAsMQwAADDEsMAAAxLDAMAMCwxDADAsMQwAADDEsMAAAxLDAMAMCwxDADAsMQwAADDEsMAAAxLDAMAMCwxDADAsMQwAADDEsMAAAxLDAMAMCwxDADAsMQwAADDEsMAAAxLDAMAMCwxDADAsMQwAADDEsMAAAxLDAMAMCwxDADAsMQwAADDEsMAAAxLDAMAMCwxDADAsMQwAADDEsMAAAxLDAMAMCwxDADAsMQwAADDEsMAAAxLDAMAMCwxDADAsMQwAADDEsMAAAxLDAMAMCwxDADAsMQwAADDEsMAAAxLDAMAMCwxDADAsLaM4ap6a1U9WFV3b7L/R6rqg1X1l1X1Z1X17NlPEwAAZm87V4ZvTHLZafZ/LMm3d/ezkvxikiMzmBcAAMzdvq0GdPftVXX4NPv/bOrmHUkOzmBeAAAwd7N+zfA1SX5/xscEAIC52PLK8HZV1T/PWgz/09OMuTbJtUly6NChWT00AACclZlcGa6qb07yliRXdvdDm43r7iPdvdLdKwcOHJjFQwMAwFnbcQxX1aEkv53kx7r7vp1PCQAAFmPLl0lU1c1Jnp9kf1UdS/LqJI9Jku5+c5JfSPKkJG+sqiQ50d0r85owAADMynZ+m8TVW+x/aZKXzmxGAACwIN6BDgCAYYlhAACGJYYBABiWGAYAYFhiGACAYYlhAACGJYYBABiWGAYAYFhiGACAYYlhAACGJYYBABiWGAYAYFhiGACAYYlhAACGJYYBABiWGAYAYFhiGACAYYlhAACGJYYBABiWGAYAYFhiGACAYYlhAACGJYYBABiWGAYAYFhiGACAYYlhAACGJYYBABiWGAYAYFhiGACAYYlhAACGJYYBABiWGAYAYFhiGACAYYlhAACGJYYBABiWGAYAYFhiGACAYYlhAACGJYYBABiWGAYAYFhiGACAYYlhAACGJYYBABiWGAYAYFhiGACAYW0Zw1X11qp6sKru3mR/VdXrq+poVX2wqv7R7KcJAACzt50rwzcmuew0+78vycWTj2uTvGnn0wIAgPnbMoa7+/YknznNkCuTvK3X3JHk71XV189qggAAMC/7ZnCMC5LcP3X72GTbp2ZwbM7Sz7/j7tx05yfztCc/Pvd9+pEkydOf8oS852e+fckz277ved2f5L5PP7Ln5g0Ay3SyAV506aH84gufueW26VZIkh/71qc+Omb6mL9xxyfSk/23feCBPPyFEzn/cftyxbMv+Ipjb3S/ZHe2yEJ/gK6qrq2q1apaPX78+CIfejg33fnJfLH7y/5wT3++F5yc716bNwAs08kGuOnOT25r2/qvs9Njprf11OcPf+FEkuThL5zY8Ngb3S/ZnV/TZxHDDyS5cOr2wcm2r9DdR7p7pbtXDhw4MIOHZjMvuvRQzqvK05/yhEe3TX++F5yc716bNwAs08kGeNGlh7a1bf3X2ekx09tq6vPzH7f24oLzH7dvw2NvdL9kd35Nr+7eelDV4STv7O5nbrDvBUlenuTyJJcmeX13P3erY66srPTq6uoZTxgAAM5EVd3V3Ssb7dvyNcNVdXOS5yfZX1XHkrw6yWOSpLvfnORdWQvho0n+Osm/ms20AQBgvraM4e6+eov9neS6mc0IAAAWxDvQAQAwrFn8ajV2k+vPX/YM5uP6h5c9AwA4t8yrGfbY12xXhgEAGJYrw+eaPfavMQBgSTRDEleGAQAYmBgGAGBYYhgAgGEN95rhw6/8vZkf8+O/9IKZH3Mj85j7PCzq/wcAjGQvN8xu5sowAADDqrU3kFu8lZWVXl1dXcpjAwAwjqq6q7tXNtrnyjAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwrH3LnsBeVlXLnsKmunvLMdud/1bH2ug423l8AGA51n/tHvnrtivDAAAMy5XhHdjr/4qa1fz3+v8HABiNr92nuDIMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwthXDVXVZVX24qo5W1Ss32H+oqt5bVX9RVR+sqstnP1UAAJitLWO4qs5L8oYk35fkGUmurqpnrBv2c0lu7e5vSXJVkjfOeqIAADBr27ky/NwkR7v7o939t0luSXLlujGd5ImTz89P8r9nN0UAAJiPfdsYc0GS+6duH0ty6box1yd5T1W9Isnjk3zXTGYHAABzNKsfoLs6yY3dfTDJ5Ul+vaq+4thVdW1VrVbV6vHjx2f00AAAcHa2E8MPJLlw6vbBybZp1yS5NUm6+8+TfE2S/esP1N1Hunulu1cOHDhwdjMGAIAZ2U4Mvy/JxVV1UVU9Nms/IHfbujGfTPKdSVJV35i1GHbpFwCAXW3LGO7uE0lenuTdST6Utd8acU9VvaaqrpgM+3dJXlZVH0hyc5KXdHfPa9IAADAL2/kBunT3u5K8a922X5j6/N4kz5vt1AAAYL68Ax0AAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADGtbMVxVl1XVh6vqaFW9cpMxP1xV91bVPVV102ynCQAAs7dvqwFVdV6SNyT57iTHkryvqm7r7nunxlyc5FVJntfdn62qJ89rwgAAMCvbuTL83CRHu/uj3f23SW5JcuW6MS9L8obu/mySdPeDs50mAADM3nZi+IIk90/dPjbZNu3pSZ5eVf+jqu6oqstmNUEAAJiXLV8mcQbHuTjJ85McTHJ7VT2ruz83Paiqrk1ybZIcOnRoRg8NAABnZztXhh9IcuHU7YOTbdOOJbmtu/+uuz+W5L6sxfGX6e4j3b3S3SsHDhw42zkDAMBMbCeG35fk4qq6qKoem+SqJLetG/OOrF0VTlXtz9rLJj46u2kCAMDsbRnD3X0iycuTvDvJh5Lc2t33VNVrquqKybB3J3moqu5N8t4k/6G7H5rXpAEAYBaqu5fywCsrK726urqUxwYAYBxVdVd3r2y0zzvQAQAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwqruX88BVx5N8YikPfu7bn+Svlj0JrMMuYR2WzxrsDtZhd7AOy/HU7j6w0Y6lxTDzU1Wr3b2y7HmMzjrsDtZh+azB7mAddgfrsPt4mQQAAMMSwwAADEsMn5uOLHsCJLEOu4V1WD5rsDtYh93BOuwyXjMMAMCwXBkGAGBYYniXqqrzquovquqdG+z7iar6y6p6f1X9aVU9Y2rfq6rqaFV9uKq+d2r7T1XV3VV1T1X99NT2v19Vf1hVH5n89+vm/uT2kAWuw/VV9cDkWO+vqsvn/uT2kLNZh6p6UlW9t6oeqapfWXef50zuc7SqXl9VNdnufDiNBa6D82ETc1iD/1RV91fVI+u2f3VV/eZkbe6sqsNzfWJ7zALX4SVVdXzqXHjpfJ/ZmMTw7vVTST60yb6buvtZ3X1Jktcm+S9JMjnhrkryTUkuS/LGyQn7zCQvS/LcJM9O8v1V9Q2TY70yyR9198VJ/mhym1MWtQ5J8rruvmTy8a75PJ0964zXIcn/S/LzSf79Bvd5U9bW4uLJx2WT7c6H01vUOiTOh83Meg1+N2t/J613TZLPdvc3JHldkl/eyaTPQYtahyT5zalz4S07mDObEMO7UFUdTPKCJBv+oe/uz0/dfHySky/8vjLJLd39N939sSRHs3ZyfWOSO7v7r7v7RJI/SfIvpu7za5PPfy3JC2f4VPa0Ba8Dmzjbdeju/9vdf5q1L0DTx/v6JE/s7jt67Ycm3pZTf+6dD5tY8DqwgVmvwWTfHd39qQ0ON30uvD3Jd568cj+6Ba8DCyCGd6cbkvzHJF/abEBVXVdV/ytr/+r8ycnmC5LcPzXs2GTb3Un+2eRbNF+b5PIkF07GPGXqBPw/SZ4yqydxDrghi1uHJHl5VX2wqt7q2/Nf5oac3Tps5oKsrclJJ9cncT6czg1Z3DokzoeN3JDZrsHpPPr32OQf7w8nedIOjncuuSGLW4ck+ZeTc+HtVXXh1sM5U2J4l6mq70/yYHffdbpx3f2G7n5akp9N8nNbjP1Q1r7F9Z4kf5Dk/Um+uMG4zqmrm0Nbwjq8KcnTklyS5FNJ/vMOpn/OmMc6bJfz4ZQlrIPzYZ1lngucsoR1+N0kh7v7m5P8YU5drWeGxPDu87wkV1TVx5PckuQ7quo3TjP+lpz61uID+fIrjQcn29Ldv9rdz+nub0vy2ST3TcZ8evLtypPftnxwRs9jr1voOnT3p7v7i939pST/NZu/dmw0O1mHzTyQtTU56dH1ifNhMwtdB+fDhuaxBqfz6N9jVbUvyflJHtrB8c4VC12H7n6ou/9mcvMtSZ5ztsdic2J4l+nuV3X3we4+nLUfwvrj7v7R6TFVdfHUzRck+cjk89uSXDX5KeCLsvYDKf9zcp8nT/57KGuvU71p6j4vnnz+4iS/M/MntQcteh1OBtjED2TtJRXD2+E6bHbMTyX5fFV96+Q1kD+eU3/unQ8bWPQ6OB++0jzWYAvT58IPTh5v+O+ULHod1p0LV2TzH9pjB/YtewJsT1W9Jslqd9+WtdfSfVeSv8va1cUXJ0l331NVtya5N8mJJNd198lvw/9WVT1pcp/ruvtzk+2/lOTWqromySeS/PCintNeNMd1eG1VXZK1b8t/PMm/Wcwz2pu2sw6TcR9P8sQkj62qFyb5nu6+N8m/TXJjkscl+f3JR+J8OCNzXAfnwzbtdA2q6rVJXpTka6vqWJK3dPf1SX41ya9X1dEkn8la+LGJOa7DT1bVFVn7WvKZJC9Z2JMaiHegAwBgWF4mAQDAsMQwAADDEsMAAAxLDAMAMCwxDADArjR5F8oHq2rLX7FYVa+rqvdPPu6rqs9t6zH8NgkAAHajqvq2JI8keVt3P/MM7veKJN/S3f96q7GuDAMAsCt19+1Z+x3Lj6qqp1XVH1TVXVX136vqH25w16uT3Lydx/CmGwAA7CVHkvxEd3+kqi5N8sYk33FyZ1U9NclFSf54OwcTwwAA7AlV9YQk/yTJf1t7J/ckyVevG3ZVkrdPvfvraYlhAAD2iq9K8rnuvuQ0Y65Kct2ZHBAAAHa97v58ko9V1Q8lSa159sn9k9cPf12SP9/uMcUwAAC7UlXdnLWw/QdVdayqrknyI0muqaoPJLknyZVTd7kqyS19Br8uza9WAwBgWK4MAwAwLDEMAMCwxDAAAMMSwwAADEsMAwAwLDEMAMCwxDAAAMMSwwAADOv/AwzUMARkWKlSAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 864x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "gene = 'A4GALT'\n",
    "df_plot = df_snp.loc[df_snp['GENE']==gene]\n",
    "\n",
    "plt.figure(figsize=[12, 5])\n",
    "plt.scatter(df_plot['BP'], [1]*df_plot.shape[0], s=4)\n",
    "for ind in df_exon.index[df_exon['GENE_NAME']==gene]:\n",
    "    plt.plot([df_exon.loc[ind, 'START'], df_exon.loc[ind, 'END']], [0.9, 0.9], color='k', linewidth=2)\n",
    "    plt.plot([df_exon.loc[ind, 'START']-1000, df_exon.loc[ind, 'START']], [0.95, 0.95], color='C0', linewidth=2)\n",
    "    plt.plot([df_exon.loc[ind, 'END'], df_exon.loc[ind, 'END']+1000], [0.975, 0.975], \n",
    "             color='C1', linewidth=2)\n",
    "plt.ylim([0.5, 1.5])\n",
    "# plt.xlim([4.3115e7,4.3119e7])\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>START</th>\n",
       "      <th>END</th>\n",
       "      <th>ENSE</th>\n",
       "      <th>ENSG</th>\n",
       "      <th>GENE_TYPE</th>\n",
       "      <th>GENE_NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>296156</th>\n",
       "      <td>chr22</td>\n",
       "      <td>43088127</td>\n",
       "      <td>43090003</td>\n",
       "      <td>ENSE00001488045</td>\n",
       "      <td>ENSG00000128274</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>A4GALT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296157</th>\n",
       "      <td>chr22</td>\n",
       "      <td>43090549</td>\n",
       "      <td>43090992</td>\n",
       "      <td>ENSE00001557395</td>\n",
       "      <td>ENSG00000128274</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>A4GALT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296158</th>\n",
       "      <td>chr22</td>\n",
       "      <td>43091497</td>\n",
       "      <td>43091637</td>\n",
       "      <td>ENSE00003825309</td>\n",
       "      <td>ENSG00000128274</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>A4GALT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296159</th>\n",
       "      <td>chr22</td>\n",
       "      <td>43116803</td>\n",
       "      <td>43116876</td>\n",
       "      <td>ENSE00003820895</td>\n",
       "      <td>ENSG00000128274</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>A4GALT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296160</th>\n",
       "      <td>chr22</td>\n",
       "      <td>43088128</td>\n",
       "      <td>43090003</td>\n",
       "      <td>ENSE00000880583</td>\n",
       "      <td>ENSG00000128274</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>A4GALT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296161</th>\n",
       "      <td>chr22</td>\n",
       "      <td>43091497</td>\n",
       "      <td>43091639</td>\n",
       "      <td>ENSE00001109573</td>\n",
       "      <td>ENSG00000128274</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>A4GALT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296162</th>\n",
       "      <td>chr22</td>\n",
       "      <td>43091581</td>\n",
       "      <td>43091637</td>\n",
       "      <td>ENSE00001559732</td>\n",
       "      <td>ENSG00000128274</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>A4GALT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296163</th>\n",
       "      <td>chr22</td>\n",
       "      <td>43116803</td>\n",
       "      <td>43116825</td>\n",
       "      <td>ENSE00003818792</td>\n",
       "      <td>ENSG00000128274</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>A4GALT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHR     START       END             ENSE             ENSG  \\\n",
       "296156  chr22  43088127  43090003  ENSE00001488045  ENSG00000128274   \n",
       "296157  chr22  43090549  43090992  ENSE00001557395  ENSG00000128274   \n",
       "296158  chr22  43091497  43091637  ENSE00003825309  ENSG00000128274   \n",
       "296159  chr22  43116803  43116876  ENSE00003820895  ENSG00000128274   \n",
       "296160  chr22  43088128  43090003  ENSE00000880583  ENSG00000128274   \n",
       "296161  chr22  43091497  43091639  ENSE00001109573  ENSG00000128274   \n",
       "296162  chr22  43091581  43091637  ENSE00001559732  ENSG00000128274   \n",
       "296163  chr22  43116803  43116825  ENSE00003818792  ENSG00000128274   \n",
       "\n",
       "             GENE_TYPE GENE_NAME  \n",
       "296156  protein_coding    A4GALT  \n",
       "296157  protein_coding    A4GALT  \n",
       "296158  protein_coding    A4GALT  \n",
       "296159  protein_coding    A4GALT  \n",
       "296160  protein_coding    A4GALT  \n",
       "296161  protein_coding    A4GALT  \n",
       "296162  protein_coding    A4GALT  \n",
       "296163  protein_coding    A4GALT  "
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_exon.loc[df_exon['GENE_NAME']==gene]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A4GALT\n",
      "AP1B1\n",
      "ARHGAP8\n",
      "ARVCF\n",
      "ASCC2\n",
      "ATP6V1E1\n",
      "ATXN10\n",
      "BCL2L13\n",
      "BCR\n",
      "BRD1\n",
      "CABIN1\n",
      "CACNA1I\n",
      "CACNG2\n",
      "CDC45\n",
      "CECR2\n",
      "CELSR1\n",
      "CERK\n",
      "CHEK2\n",
      "CLTCL1\n",
      "CSNK1E\n",
      "DEPDC5\n",
      "DGCR2\n",
      "DGCR5\n",
      "DGCR8\n",
      "EFCAB6\n",
      "EIF4ENIF1\n",
      "EMID1\n",
      "FAM118A\n",
      "FAM19A5\n",
      "FBLN1\n",
      "GGT1\n",
      "GRAMD4\n",
      "GRAP2\n",
      "GTPBP1\n",
      "GTSE1\n",
      "IL2RB\n",
      "KIAA0930\n",
      "KIAA1671\n",
      "KLHL22\n",
      "LARGE\n",
      "LIMK2\n",
      "LRP5L\n",
      "MED15\n",
      "MEI1\n",
      "MICAL3\n",
      "MICALL1\n",
      "MKL1\n",
      "MORC2\n",
      "MOV10L1\n",
      "MTMR3\n",
      "MYH9\n",
      "MYO18B\n",
      "NDUFA6-AS1\n",
      "OSBP2\n",
      "PACSIN2\n",
      "PARVB\n",
      "PI4KA\n",
      "PI4KAP2\n",
      "PISD\n",
      "PITPNB\n",
      "PLA2G6\n",
      "PLXNB2\n",
      "POLR2F\n",
      "PPARA\n",
      "PPIL2\n",
      "PPP6R2\n",
      "PRR5\n",
      "RBFOX2\n",
      "RFPL1S\n",
      "RNF215\n",
      "SCUBE1\n",
      "SEZ6L\n",
      "SFI1\n",
      "SGSM1\n",
      "SMTN\n",
      "SPECC1L\n",
      "SREBF2\n",
      "SYN3\n",
      "TAB1\n",
      "TANGO2\n",
      "TBC1D10A\n",
      "TCF20\n",
      "TMEM184B\n",
      "TNRC6B\n",
      "TPST2\n",
      "TRIOBP\n",
      "TTC28\n",
      "TTC28-AS1\n",
      "TTC38\n",
      "TXNRD2\n",
      "UPB1\n",
      "WBP2NL\n",
      "ZNRF3\n"
     ]
    }
   ],
   "source": [
    "temp_df = df_snp.groupby('GENE').agg({'BP':list})  \n",
    "for gene in temp_df.index:\n",
    "    if np.std(temp_df.loc[gene, 'BP']) > 10000:\n",
    "        print(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BP</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>[43118163, 43092620, 43090118, 43116929, 43116...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACO2</th>\n",
       "      <td>[41864700, 41865079, 41865809, 41865739, 41865...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACR</th>\n",
       "      <td>[51176855, 51177420, 51176616, 51176004, 51175...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADM2</th>\n",
       "      <td>[50919900, 50919184, 50920618, 50919439, 50919...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADORA2A</th>\n",
       "      <td>[24814772, 24827659, 24820280, 24820094, 24814...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF280B</th>\n",
       "      <td>[22862546, 22864105, 22863768, 22862813, 22864...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF70</th>\n",
       "      <td>[24093747, 24094158, 24092758, 24094109, 24093...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF74</th>\n",
       "      <td>[20749335, 20748053, 20749042, 20749298, 20747...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNRF3</th>\n",
       "      <td>[29313256, 29280436, 29278720, 29312578, 29279...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNRF3-AS1</th>\n",
       "      <td>[29427376, 29427200, 29428269, 29427337, 29427...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>459 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                          BP\n",
       "GENE                                                        \n",
       "A4GALT     [43118163, 43092620, 43090118, 43116929, 43116...\n",
       "ACO2       [41864700, 41865079, 41865809, 41865739, 41865...\n",
       "ACR        [51176855, 51177420, 51176616, 51176004, 51175...\n",
       "ADM2       [50919900, 50919184, 50920618, 50919439, 50919...\n",
       "ADORA2A    [24814772, 24827659, 24820280, 24820094, 24814...\n",
       "...                                                      ...\n",
       "ZNF280B    [22862546, 22864105, 22863768, 22862813, 22864...\n",
       "ZNF70      [24093747, 24094158, 24092758, 24094109, 24093...\n",
       "ZNF74      [20749335, 20748053, 20749042, 20749298, 20747...\n",
       "ZNRF3      [29313256, 29280436, 29278720, 29312578, 29279...\n",
       "ZNRF3-AS1  [29427376, 29427200, 29428269, 29427337, 29427...\n",
       "\n",
       "[459 rows x 1 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Old code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_info = pd.DataFrame(index=DIC_NAME_pAN)\n",
    "df_info['Name'] = [DIC_NAME_pAN[x][0] for x in df_info.index]\n",
    "df_info['Identifier'] = df_info.index\n",
    "df_info['Type'] = [DIC_NAME_pAN[x][2] for x in df_info.index]\n",
    "df_info['n_pair_common_common'] = 0\n",
    "df_info['n_pair_common_lf'] = 0\n",
    "df_info['n_pair_lf_lf'] = 0\n",
    "df_info['median_dist_common_common'] = 0\n",
    "df_info['median_dist_common_lf'] = 0\n",
    "df_info['median_dist_lf_lf'] = 0\n",
    "df_info['avgr_common_common'] = 0\n",
    "df_info['avgr_common_lf'] = 0\n",
    "df_info['avgr_lf_lf'] = 0\n",
    "df_info['Description'] = [DIC_NAME_pAN[x][1] for x in df_info.index]\n",
    "\n",
    "# n_pair & median distance\n",
    "STATS_FILE = '/n/groups/price/martin/data_GDREG/UKBimp_337K_MAF001/pannot/stats/@.dist.tsv'\n",
    "for row in df_info.index:\n",
    "    for mbin in ['common_common', 'common_lf', 'lf_lf']:\n",
    "        if os.path.exists(STATS_FILE.replace('@', '%s_%s' % (row,mbin))):\n",
    "            temp_df = pd.read_csv(STATS_FILE.replace('@', '%s_%s' % (row,mbin)), sep='\\t')\n",
    "            df_info.loc[row, 'n_pair_%s' % mbin] = temp_df['count'].sum()\n",
    "for row in df_info.index[df_info['Type']=='fct-100']:\n",
    "    for mbin in ['common_common', 'lf_lf']:\n",
    "        df_info.loc[row, 'n_pair_%s'%mbin] = \\\n",
    "            dic_res['fct100']['summary']['rho'].loc['pAN:%s_%s'%(row,mbin), 'n_pair']\n",
    "for row in df_info.index[df_info['Type']=='fct-1k']:\n",
    "    for mbin in ['common_common', 'lf_lf']:\n",
    "        df_info.loc[row, 'n_pair_%s'%mbin] = \\\n",
    "            dic_res['fct1k']['summary']['rho'].loc['pAN:%s_%s'%(row,mbin), 'n_pair']\n",
    "            \n",
    "# avgr \n",
    "AVGR_FILE = '/n/groups/price/martin/data_GDREG/UKBimp_337K_MAF001/pannot/main.avgr'\n",
    "temp_df = pd.read_csv(AVGR_FILE, sep='\\t', header=None)\n",
    "temp_df.index = temp_df[0]\n",
    "for row in df_info.index:\n",
    "    for mbin in ['common_common', 'common_lf', 'lf_lf']:\n",
    "        df_info.loc[row, 'avgr_%s' % mbin] = temp_df.loc['pAN:%s_%s' % (row, mbin), 1]\n",
    "            \n",
    "df_info.to_csv('/n/groups/price/martin/data_GDREG/results/tables/pannot_info.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n_pannot=32\n",
      "\\begin{tabular}{llllll}\n",
      "\\toprule\n",
      "                      Name &                        Description &  Size (C) & Size (LF) & Avg. LD (C) & Avg. LD (LF) \\\\\n",
      "\\midrule\n",
      "          Proximal 0-100bp &                            <100 bp &     2.62M &     0.98M &       0.416 &        0.309 \\\\\n",
      "            Proximal 0-1kb &                              <1 kb &    23.62M &     8.46M &       0.364 &        0.249 \\\\\n",
      "        Proximal 100bp-1kb &                          100bp-1kb &    21.00M &     7.48M &       0.358 &        0.241 \\\\\n",
      "                      Exon &                          Same exon &     0.54M &     0.26M &       0.278 &        0.138 \\\\\n",
      "                      Gene &                          Same gene &  1367.49M &   521.53M &       0.065 &        0.047 \\\\\n",
      "               Exonic-gene &         Exonic region of same gene &     1.17M &     0.61M &       0.205 &        0.098 \\\\\n",
      "            Protein-domain &  Same protein domain and same gene &     0.09M &     0.09M &       0.206 &        0.073 \\\\\n",
      "                  Promoter &              Promoter of same gene &     0.88M &     0.37M &       0.218 &        0.133 \\\\\n",
      "       H3K27ac (Hnisz)-100 &           H3K27ac (Hnisz) (<100bp) &     0.99M &     0.38M &       0.396 &        0.304 \\\\\n",
      "        H3K27ac (PGC2)-100 &            H3K27ac (PGC2) (<100bp) &     0.66M &     0.26M &       0.386 &        0.301 \\\\\n",
      "               H3K4me1-100 &                   H3K4me1 (<100bp) &     0.99M &     0.38M &       0.400 &        0.294 \\\\\n",
      "                Intron-100 &                    Intron (<100bp) &     0.94M &     0.38M &       0.416 &        0.295 \\\\\n",
      "             Repressed-100 &                 Repressed (<100bp) &     1.16M &     0.41M &       0.425 &        0.325 \\\\\n",
      "           Transcribed-100 &               Transcribed (<100bp) &     0.75M &     0.31M &       0.425 &        0.296 \\\\\n",
      "                    DGF-1k &                         DGF (<1kb) &     0.85M &     0.34M &       0.348 &        0.243 \\\\\n",
      "                    DHS-1k &                         DHS (<1kb) &     1.38M &     0.53M &       0.355 &        0.236 \\\\\n",
      "              DHS peaks-1k &                   DHS peaks (<1kb) &     0.65M &     0.26M &       0.354 &        0.233 \\\\\n",
      "               Enhancer-1k &                    Enhancer (<1kb) &     0.52M &     0.20M &       0.351 &        0.231 \\\\\n",
      "              Fetal DHS-1k &                   Fetal DHS (<1kb) &     0.60M &     0.24M &       0.344 &        0.228 \\\\\n",
      "        H3K27ac (Hnisz)-1k &             H3K27ac (Hnisz) (<1kb) &     8.53M &     3.11M &       0.337 &        0.242 \\\\\n",
      "         H3K27ac (PGC2)-1k &              H3K27ac (PGC2) (<1kb) &     5.29M &     1.96M &       0.327 &        0.235 \\\\\n",
      "                H3K4me1-1k &                     H3K4me1 (<1kb) &     7.47M &     2.79M &       0.342 &        0.234 \\\\\n",
      "          H3K4me1 peaks-1k &               H3K4me1 peaks (<1kb) &     1.49M &     0.60M &       0.345 &        0.226 \\\\\n",
      "                H3K4me3-1k &                     H3K4me3 (<1kb) &     1.92M &     0.77M &       0.327 &        0.216 \\\\\n",
      "                 H3K9ac-1k &                      H3K9ac (<1kb) &     1.76M &     0.72M &       0.359 &        0.217 \\\\\n",
      "                 Intron-1k &                      Intron (<1kb) &     8.29M &     3.18M &       0.362 &        0.231 \\\\\n",
      "               Promoter-1k &                    Promoter (<1kb) &     0.95M &     0.37M &       0.321 &        0.206 \\\\\n",
      "              Repressed-1k &                   Repressed (<1kb) &     8.20M &     2.76M &       0.375 &        0.268 \\\\\n",
      " Super enhancer (Hnisz)-1k &      Super enhancer (Hnisz) (<1kb) &     3.89M &     1.40M &       0.306 &        0.246 \\\\\n",
      "                   TFBS-1k &                        TFBS (<1kb) &     1.40M &     0.51M &       0.306 &        0.232 \\\\\n",
      "            Transcribed-1k &                 Transcribed (<1kb) &     4.85M &     1.97M &       0.374 &        0.223 \\\\\n",
      "       Typical enhancer-1k &            Typical enhancer (<1kb) &     0.48M &     0.17M &       0.255 &        0.270 \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Table 1 pannots\n",
    "DF_PANNOT = pd.read_csv('/n/groups/price/martin/data_GDREG/results/tables/pannot_info.tsv', sep='\\t')\n",
    "# DF_PANNOT.index = DF_PANNOT['Identifier']\n",
    "print(\"n_pannot=%d\" % DF_PANNOT.shape[0])\n",
    "DF_PANNOT = DF_PANNOT[['Name', 'Description', 'n_pair_common_common', 'n_pair_lf_lf', \n",
    "                       'avgr_common_common', 'avgr_lf_lf']]\n",
    "for col in ['n_pair_common_common', 'n_pair_lf_lf']:\n",
    "    DF_PANNOT[col] = ['%0.2fM' % (x/1e6) for x in DF_PANNOT[col]]\n",
    "for col in ['avgr_common_common', 'avgr_lf_lf']:\n",
    "    DF_PANNOT[col] = ['%0.3f' % x for x in DF_PANNOT[col]]\n",
    "DF_PANNOT.columns = ['Name', 'Description', 'Size (C)', 'Size (LF)', 'Avg. LD (C)', 'Avg. LD (LF)']\n",
    "print(DF_PANNOT.to_latex(index=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Table 1 pannots\n",
    "DF_PANNOT = pd.read_csv('/n/groups/price/martin/data_GDREG/results/tables/pannot_info.tsv', sep='\\t')\n",
    "# DF_PANNOT.index = DF_PANNOT['Identifier']\n",
    "print(\"n_pannot=%d\" % DF_PANNOT.shape[0])\n",
    "DF_PANNOT = DF_PANNOT[['Name', 'Description', 'n_pair_common_common', 'n_pair_lf_lf', \n",
    "                       'avgr_common_common', 'avgr_lf_lf']]\n",
    "for col in ['n_pair_common_common', 'n_pair_lf_lf']:\n",
    "    DF_PANNOT[col] = ['%0.2fM' % (x/1e6) for x in DF_PANNOT[col]]\n",
    "for col in ['avgr_common_common', 'avgr_lf_lf']:\n",
    "    DF_PANNOT[col] = ['%0.3f' % x for x in DF_PANNOT[col]]\n",
    "DF_PANNOT.columns = ['Name', 'Description', 'Size (C)', 'Size (LF)', 'Avg. LD (C)', 'Avg. LD (LF)']\n",
    "print(DF_PANNOT.to_latex(index=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n_pannot=30\n",
      "\\begin{tabular}{llllll}\n",
      "\\toprule\n",
      "                      Name &                        Description &  Size (C) & Size (LF) & Avg. LD (C) & Avg. LD (LF) \\\\\n",
      "\\midrule\n",
      "          Proximal 0-100bp &                            <100 bp &     2.62M &     0.98M &       0.416 &        0.309 \\\\\n",
      "            Proximal 0-1kb &                              <1 kb &    23.62M &     8.46M &       0.364 &        0.249 \\\\\n",
      "        Proximal 100bp-1kb &                          100bp-1kb &    21.00M &     7.48M &       0.358 &        0.241 \\\\\n",
      "                      Exon &                          Same exon &     0.54M &     0.26M &       0.278 &        0.138 \\\\\n",
      "                      Gene &                          Same gene &  1367.49M &   521.53M &       0.065 &        0.047 \\\\\n",
      "               Exonic-gene &         Exonic region of same gene &     1.17M &     0.61M &       0.205 &        0.098 \\\\\n",
      "            Protein-domain &  Same protein domain and same gene &     0.09M &     0.09M &       0.206 &        0.073 \\\\\n",
      "                  Promoter &              Promoter of same gene &     0.88M &     0.37M &       0.218 &        0.133 \\\\\n",
      "       H3K27ac (Hnisz)-100 &           H3K27ac (Hnisz) (<100bp) &     1.98M &     0.76M &       0.396 &        0.304 \\\\\n",
      "        H3K27ac (PGC2)-100 &            H3K27ac (PGC2) (<100bp) &     1.33M &     0.52M &       0.386 &        0.301 \\\\\n",
      "               H3K4me1-100 &                   H3K4me1 (<100bp) &     1.97M &     0.76M &       0.400 &        0.294 \\\\\n",
      "                Intron-100 &                    Intron (<100bp) &     1.88M &     0.75M &       0.416 &        0.295 \\\\\n",
      "             Repressed-100 &                 Repressed (<100bp) &     2.31M &     0.82M &       0.425 &        0.325 \\\\\n",
      "           Transcribed-100 &               Transcribed (<100bp) &     1.51M &     0.62M &       0.425 &        0.296 \\\\\n",
      "                    DGF-1k &                         DGF (<1kb) &     1.70M &     0.68M &       0.348 &        0.243 \\\\\n",
      "                    DHS-1k &                         DHS (<1kb) &     2.76M &     1.07M &       0.355 &        0.236 \\\\\n",
      "               Enhancer-1k &                    Enhancer (<1kb) &     1.05M &     0.40M &       0.351 &        0.231 \\\\\n",
      "              Fetal DHS-1k &                   Fetal DHS (<1kb) &     1.20M &     0.48M &       0.344 &        0.228 \\\\\n",
      "        H3K27ac (Hnisz)-1k &             H3K27ac (Hnisz) (<1kb) &    17.05M &     6.23M &       0.337 &        0.242 \\\\\n",
      "         H3K27ac (PGC2)-1k &              H3K27ac (PGC2) (<1kb) &    10.57M &     3.93M &       0.327 &        0.235 \\\\\n",
      "                H3K4me1-1k &                     H3K4me1 (<1kb) &    14.95M &     5.59M &       0.342 &        0.234 \\\\\n",
      "                H3K4me3-1k &                     H3K4me3 (<1kb) &     3.83M &     1.55M &       0.327 &        0.216 \\\\\n",
      "                 H3K9ac-1k &                      H3K9ac (<1kb) &     3.52M &     1.44M &       0.359 &        0.217 \\\\\n",
      "                 Intron-1k &                      Intron (<1kb) &    16.58M &     6.37M &       0.362 &        0.231 \\\\\n",
      "               Promoter-1k &                    Promoter (<1kb) &     1.90M &     0.73M &       0.321 &        0.206 \\\\\n",
      "              Repressed-1k &                   Repressed (<1kb) &    16.40M &     5.52M &       0.375 &        0.268 \\\\\n",
      " Super enhancer (Hnisz)-1k &      Super enhancer (Hnisz) (<1kb) &     7.78M &     2.80M &       0.306 &        0.246 \\\\\n",
      "                   TFBS-1k &                        TFBS (<1kb) &     2.80M &     1.02M &       0.306 &        0.232 \\\\\n",
      "            Transcribed-1k &                 Transcribed (<1kb) &     9.70M &     3.94M &       0.374 &        0.223 \\\\\n",
      "       Typical enhancer-1k &            Typical enhancer (<1kb) &     0.96M &     0.34M &       0.255 &        0.270 \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "DF_PANNOT = pd.read_csv('/n/groups/price/martin/data_GDREG/results/tables/pannot_info.tsv', sep='\\t')\n",
    "# DF_PANNOT.index = DF_PANNOT['Identifier']\n",
    "print(\"n_pannot=%d\" % DF_PANNOT.shape[0])\n",
    "DF_PANNOT = DF_PANNOT[['Name', 'Description', 'n_pair_common_common', 'n_pair_lf_lf', \n",
    "                       'avgr_common_common', 'avgr_lf_lf']]\n",
    "for col in ['n_pair_common_common', 'n_pair_lf_lf']:\n",
    "    DF_PANNOT[col] = ['%0.2fM' % (x/1e6) for x in DF_PANNOT[col]]\n",
    "for col in ['avgr_common_common', 'avgr_lf_lf']:\n",
    "    DF_PANNOT[col] = ['%0.3f' % x for x in DF_PANNOT[col]]\n",
    "DF_PANNOT.columns = ['Name', 'Description', 'Size (C)', 'Size (LF)', 'Avg. LD (C)', 'Avg. LD (LF)']\n",
    "print(DF_PANNOT.to_latex(index=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Description</th>\n",
       "      <th>Type</th>\n",
       "      <th>Size (C)</th>\n",
       "      <th>Size (LF)</th>\n",
       "      <th>Avg. LD (C)</th>\n",
       "      <th>Avg. LD (LF)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Proximal (0-0.1 kb)</td>\n",
       "      <td>Within 100 bp</td>\n",
       "      <td>proximity</td>\n",
       "      <td>2621179</td>\n",
       "      <td>981186</td>\n",
       "      <td>0.416440</td>\n",
       "      <td>0.309336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Proximal (0-1 kb)</td>\n",
       "      <td>Within 1 kb</td>\n",
       "      <td>proximity</td>\n",
       "      <td>23622561</td>\n",
       "      <td>8458072</td>\n",
       "      <td>0.364336</td>\n",
       "      <td>0.248938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Proximal (0.1-1 kb)</td>\n",
       "      <td>Between 100 bp and 1 kb</td>\n",
       "      <td>proximity</td>\n",
       "      <td>21001382</td>\n",
       "      <td>7476886</td>\n",
       "      <td>0.357821</td>\n",
       "      <td>0.240968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Exon</td>\n",
       "      <td>Same exon</td>\n",
       "      <td>gene</td>\n",
       "      <td>542754</td>\n",
       "      <td>262898</td>\n",
       "      <td>0.278384</td>\n",
       "      <td>0.137887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Same gene</td>\n",
       "      <td>gene</td>\n",
       "      <td>1367493711</td>\n",
       "      <td>521528493</td>\n",
       "      <td>0.065348</td>\n",
       "      <td>0.046902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Exonic-gene</td>\n",
       "      <td>Exonic region of the same gene</td>\n",
       "      <td>gene</td>\n",
       "      <td>1173399</td>\n",
       "      <td>608234</td>\n",
       "      <td>0.204768</td>\n",
       "      <td>0.097622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Protein-domain</td>\n",
       "      <td>Same protein domain and same gene</td>\n",
       "      <td>gene</td>\n",
       "      <td>93951</td>\n",
       "      <td>91867</td>\n",
       "      <td>0.205562</td>\n",
       "      <td>0.072658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Promoter</td>\n",
       "      <td>Pairs of promoter SNPs linked to the same gene</td>\n",
       "      <td>gene</td>\n",
       "      <td>877113</td>\n",
       "      <td>365052</td>\n",
       "      <td>0.217566</td>\n",
       "      <td>0.132904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>H3K27ac (Hnisz)-100</td>\n",
       "      <td>SNP pairs within 100 bp for H3K27ac (Hnisz)</td>\n",
       "      <td>fct-100</td>\n",
       "      <td>1975114</td>\n",
       "      <td>756672</td>\n",
       "      <td>0.396433</td>\n",
       "      <td>0.304018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>H3K27ac (PGC2)-100</td>\n",
       "      <td>SNP pairs within 100 bp for H3K27ac (PGC2)</td>\n",
       "      <td>fct-100</td>\n",
       "      <td>1328988</td>\n",
       "      <td>517016</td>\n",
       "      <td>0.386332</td>\n",
       "      <td>0.301221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>H3K4me1-100</td>\n",
       "      <td>SNP pairs within 100 bp for H3K4me1</td>\n",
       "      <td>fct-100</td>\n",
       "      <td>1971744</td>\n",
       "      <td>763964</td>\n",
       "      <td>0.400148</td>\n",
       "      <td>0.293532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Intron-100</td>\n",
       "      <td>SNP pairs within 100 bp for Intron</td>\n",
       "      <td>fct-100</td>\n",
       "      <td>1882944</td>\n",
       "      <td>754116</td>\n",
       "      <td>0.416202</td>\n",
       "      <td>0.295014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Repressed-100</td>\n",
       "      <td>SNP pairs within 100 bp for Repressed</td>\n",
       "      <td>fct-100</td>\n",
       "      <td>2313264</td>\n",
       "      <td>816120</td>\n",
       "      <td>0.425160</td>\n",
       "      <td>0.325253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Transcribed-100</td>\n",
       "      <td>SNP pairs within 100 bp for Transcribed</td>\n",
       "      <td>fct-100</td>\n",
       "      <td>1506552</td>\n",
       "      <td>621556</td>\n",
       "      <td>0.424660</td>\n",
       "      <td>0.296087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>DGF-1k</td>\n",
       "      <td>SNP pairs within 1 kb for DGF</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>1698534</td>\n",
       "      <td>682010</td>\n",
       "      <td>0.348136</td>\n",
       "      <td>0.243331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>DHS-1k</td>\n",
       "      <td>SNP pairs within 1 kb for DHS</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>2761874</td>\n",
       "      <td>1069206</td>\n",
       "      <td>0.355067</td>\n",
       "      <td>0.235842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Enhancer-1k</td>\n",
       "      <td>SNP pairs within 1 kb for Enhancer</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>1048720</td>\n",
       "      <td>402016</td>\n",
       "      <td>0.350655</td>\n",
       "      <td>0.230790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Fetal DHS-1k</td>\n",
       "      <td>SNP pairs within 1 kb for Fetal DHS</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>1198364</td>\n",
       "      <td>483422</td>\n",
       "      <td>0.343788</td>\n",
       "      <td>0.228424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>H3K27ac (Hnisz)-1k</td>\n",
       "      <td>SNP pairs within 1 kb for H3K27ac (Hnisz)</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>17053934</td>\n",
       "      <td>6226602</td>\n",
       "      <td>0.337385</td>\n",
       "      <td>0.241750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>H3K27ac (PGC2)-1k</td>\n",
       "      <td>SNP pairs within 1 kb for H3K27ac (PGC2)</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>10573720</td>\n",
       "      <td>3927250</td>\n",
       "      <td>0.327117</td>\n",
       "      <td>0.235396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>H3K4me1-1k</td>\n",
       "      <td>SNP pairs within 1 kb for H3K4me1</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>14950834</td>\n",
       "      <td>5590064</td>\n",
       "      <td>0.342022</td>\n",
       "      <td>0.233561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>H3K4me3-1k</td>\n",
       "      <td>SNP pairs within 1 kb for H3K4me3</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>3833666</td>\n",
       "      <td>1547670</td>\n",
       "      <td>0.326912</td>\n",
       "      <td>0.215755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>H3K9ac-1k</td>\n",
       "      <td>SNP pairs within 1 kb for H3K9ac</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>3522624</td>\n",
       "      <td>1442666</td>\n",
       "      <td>0.358715</td>\n",
       "      <td>0.216737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Intron-1k</td>\n",
       "      <td>SNP pairs within 1 kb for Intron</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>16575870</td>\n",
       "      <td>6366986</td>\n",
       "      <td>0.361928</td>\n",
       "      <td>0.231346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Promoter-1k</td>\n",
       "      <td>SNP pairs within 1 kb for Promoter</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>1901976</td>\n",
       "      <td>732224</td>\n",
       "      <td>0.321337</td>\n",
       "      <td>0.206426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Repressed-1k</td>\n",
       "      <td>SNP pairs within 1 kb for Repressed</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>16403016</td>\n",
       "      <td>5524246</td>\n",
       "      <td>0.374997</td>\n",
       "      <td>0.268214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Super enhancer (Hnisz)-1k</td>\n",
       "      <td>SNP pairs within 1 kb for Super enhancer (Hnisz)</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>7779330</td>\n",
       "      <td>2802540</td>\n",
       "      <td>0.306412</td>\n",
       "      <td>0.245634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>TFBS-1k</td>\n",
       "      <td>SNP pairs within 1 kb for TFBS</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>2804782</td>\n",
       "      <td>1024928</td>\n",
       "      <td>0.306054</td>\n",
       "      <td>0.232053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Transcribed-1k</td>\n",
       "      <td>SNP pairs within 1 kb for Transcribed</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>9696862</td>\n",
       "      <td>3944834</td>\n",
       "      <td>0.374280</td>\n",
       "      <td>0.222523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Typical enhancer-1k</td>\n",
       "      <td>SNP pairs within 1 kb for Typical enhancer</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>961102</td>\n",
       "      <td>342576</td>\n",
       "      <td>0.255268</td>\n",
       "      <td>0.269883</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Name  \\\n",
       "0         Proximal (0-0.1 kb)   \n",
       "1           Proximal (0-1 kb)   \n",
       "2         Proximal (0.1-1 kb)   \n",
       "3                        Exon   \n",
       "4                        Gene   \n",
       "5                 Exonic-gene   \n",
       "6              Protein-domain   \n",
       "7                    Promoter   \n",
       "8         H3K27ac (Hnisz)-100   \n",
       "9          H3K27ac (PGC2)-100   \n",
       "10                H3K4me1-100   \n",
       "11                 Intron-100   \n",
       "12              Repressed-100   \n",
       "13            Transcribed-100   \n",
       "14                     DGF-1k   \n",
       "15                     DHS-1k   \n",
       "16                Enhancer-1k   \n",
       "17               Fetal DHS-1k   \n",
       "18         H3K27ac (Hnisz)-1k   \n",
       "19          H3K27ac (PGC2)-1k   \n",
       "20                 H3K4me1-1k   \n",
       "21                 H3K4me3-1k   \n",
       "22                  H3K9ac-1k   \n",
       "23                  Intron-1k   \n",
       "24                Promoter-1k   \n",
       "25               Repressed-1k   \n",
       "26  Super enhancer (Hnisz)-1k   \n",
       "27                    TFBS-1k   \n",
       "28             Transcribed-1k   \n",
       "29        Typical enhancer-1k   \n",
       "\n",
       "                                         Description       Type    Size (C)  \\\n",
       "0                                      Within 100 bp  proximity     2621179   \n",
       "1                                        Within 1 kb  proximity    23622561   \n",
       "2                            Between 100 bp and 1 kb  proximity    21001382   \n",
       "3                                          Same exon       gene      542754   \n",
       "4                                          Same gene       gene  1367493711   \n",
       "5                     Exonic region of the same gene       gene     1173399   \n",
       "6                  Same protein domain and same gene       gene       93951   \n",
       "7     Pairs of promoter SNPs linked to the same gene       gene      877113   \n",
       "8        SNP pairs within 100 bp for H3K27ac (Hnisz)    fct-100     1975114   \n",
       "9         SNP pairs within 100 bp for H3K27ac (PGC2)    fct-100     1328988   \n",
       "10               SNP pairs within 100 bp for H3K4me1    fct-100     1971744   \n",
       "11                SNP pairs within 100 bp for Intron    fct-100     1882944   \n",
       "12             SNP pairs within 100 bp for Repressed    fct-100     2313264   \n",
       "13           SNP pairs within 100 bp for Transcribed    fct-100     1506552   \n",
       "14                     SNP pairs within 1 kb for DGF     fct-1k     1698534   \n",
       "15                     SNP pairs within 1 kb for DHS     fct-1k     2761874   \n",
       "16                SNP pairs within 1 kb for Enhancer     fct-1k     1048720   \n",
       "17               SNP pairs within 1 kb for Fetal DHS     fct-1k     1198364   \n",
       "18         SNP pairs within 1 kb for H3K27ac (Hnisz)     fct-1k    17053934   \n",
       "19          SNP pairs within 1 kb for H3K27ac (PGC2)     fct-1k    10573720   \n",
       "20                 SNP pairs within 1 kb for H3K4me1     fct-1k    14950834   \n",
       "21                 SNP pairs within 1 kb for H3K4me3     fct-1k     3833666   \n",
       "22                  SNP pairs within 1 kb for H3K9ac     fct-1k     3522624   \n",
       "23                  SNP pairs within 1 kb for Intron     fct-1k    16575870   \n",
       "24                SNP pairs within 1 kb for Promoter     fct-1k     1901976   \n",
       "25               SNP pairs within 1 kb for Repressed     fct-1k    16403016   \n",
       "26  SNP pairs within 1 kb for Super enhancer (Hnisz)     fct-1k     7779330   \n",
       "27                    SNP pairs within 1 kb for TFBS     fct-1k     2804782   \n",
       "28             SNP pairs within 1 kb for Transcribed     fct-1k     9696862   \n",
       "29        SNP pairs within 1 kb for Typical enhancer     fct-1k      961102   \n",
       "\n",
       "    Size (LF)  Avg. LD (C)  Avg. LD (LF)  \n",
       "0      981186     0.416440      0.309336  \n",
       "1     8458072     0.364336      0.248938  \n",
       "2     7476886     0.357821      0.240968  \n",
       "3      262898     0.278384      0.137887  \n",
       "4   521528493     0.065348      0.046902  \n",
       "5      608234     0.204768      0.097622  \n",
       "6       91867     0.205562      0.072658  \n",
       "7      365052     0.217566      0.132904  \n",
       "8      756672     0.396433      0.304018  \n",
       "9      517016     0.386332      0.301221  \n",
       "10     763964     0.400148      0.293532  \n",
       "11     754116     0.416202      0.295014  \n",
       "12     816120     0.425160      0.325253  \n",
       "13     621556     0.424660      0.296087  \n",
       "14     682010     0.348136      0.243331  \n",
       "15    1069206     0.355067      0.235842  \n",
       "16     402016     0.350655      0.230790  \n",
       "17     483422     0.343788      0.228424  \n",
       "18    6226602     0.337385      0.241750  \n",
       "19    3927250     0.327117      0.235396  \n",
       "20    5590064     0.342022      0.233561  \n",
       "21    1547670     0.326912      0.215755  \n",
       "22    1442666     0.358715      0.216737  \n",
       "23    6366986     0.361928      0.231346  \n",
       "24     732224     0.321337      0.206426  \n",
       "25    5524246     0.374997      0.268214  \n",
       "26    2802540     0.306412      0.245634  \n",
       "27    1024928     0.306054      0.232053  \n",
       "28    3944834     0.374280      0.222523  \n",
       "29     342576     0.255268      0.269883  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_PANNOT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Type</th>\n",
       "      <th>n_pair_common_common</th>\n",
       "      <th>n_pair_lf_lf</th>\n",
       "      <th>avgr_common_common</th>\n",
       "      <th>avgr_lf_lf</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Proximal (0-0.1 kb)</td>\n",
       "      <td>proximity</td>\n",
       "      <td>2621179</td>\n",
       "      <td>981186</td>\n",
       "      <td>0.416440</td>\n",
       "      <td>0.309336</td>\n",
       "      <td>Within 100 bp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Proximal (0-1 kb)</td>\n",
       "      <td>proximity</td>\n",
       "      <td>23622561</td>\n",
       "      <td>8458072</td>\n",
       "      <td>0.364336</td>\n",
       "      <td>0.248938</td>\n",
       "      <td>Within 1 kb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Proximal (0.1-1 kb)</td>\n",
       "      <td>proximity</td>\n",
       "      <td>21001382</td>\n",
       "      <td>7476886</td>\n",
       "      <td>0.357821</td>\n",
       "      <td>0.240968</td>\n",
       "      <td>Between 100 bp and 1 kb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Exon</td>\n",
       "      <td>gene</td>\n",
       "      <td>542754</td>\n",
       "      <td>262898</td>\n",
       "      <td>0.278384</td>\n",
       "      <td>0.137887</td>\n",
       "      <td>Same exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gene</td>\n",
       "      <td>gene</td>\n",
       "      <td>1367493711</td>\n",
       "      <td>521528493</td>\n",
       "      <td>0.065348</td>\n",
       "      <td>0.046902</td>\n",
       "      <td>Same gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Exonic-gene</td>\n",
       "      <td>gene</td>\n",
       "      <td>1173399</td>\n",
       "      <td>608234</td>\n",
       "      <td>0.204768</td>\n",
       "      <td>0.097622</td>\n",
       "      <td>Exonic region of the same gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Protein-domain</td>\n",
       "      <td>gene</td>\n",
       "      <td>93951</td>\n",
       "      <td>91867</td>\n",
       "      <td>0.205562</td>\n",
       "      <td>0.072658</td>\n",
       "      <td>Same protein domain and same gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Promoter</td>\n",
       "      <td>gene</td>\n",
       "      <td>877113</td>\n",
       "      <td>365052</td>\n",
       "      <td>0.217566</td>\n",
       "      <td>0.132904</td>\n",
       "      <td>Pairs of promoter SNPs linked to the same gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>H3K27ac (Hnisz)-100</td>\n",
       "      <td>fct-100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.396433</td>\n",
       "      <td>0.304018</td>\n",
       "      <td>SNP pairs within 100 bp for H3K27ac (Hnisz)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>H3K27ac (PGC2)-100</td>\n",
       "      <td>fct-100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.386332</td>\n",
       "      <td>0.301221</td>\n",
       "      <td>SNP pairs within 100 bp for H3K27ac (PGC2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>H3K4me1-100</td>\n",
       "      <td>fct-100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.400148</td>\n",
       "      <td>0.293532</td>\n",
       "      <td>SNP pairs within 100 bp for H3K4me1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Intron-100</td>\n",
       "      <td>fct-100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.416202</td>\n",
       "      <td>0.295014</td>\n",
       "      <td>SNP pairs within 100 bp for Intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Repressed-100</td>\n",
       "      <td>fct-100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.425160</td>\n",
       "      <td>0.325253</td>\n",
       "      <td>SNP pairs within 100 bp for Repressed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Transcribed-100</td>\n",
       "      <td>fct-100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.424660</td>\n",
       "      <td>0.296087</td>\n",
       "      <td>SNP pairs within 100 bp for Transcribed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>DGF-1k</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.348136</td>\n",
       "      <td>0.243331</td>\n",
       "      <td>SNP pairs within 1 kb for DGF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>DHS-1k</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.355067</td>\n",
       "      <td>0.235842</td>\n",
       "      <td>SNP pairs within 1 kb for DHS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Enhancer-1k</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.350655</td>\n",
       "      <td>0.230790</td>\n",
       "      <td>SNP pairs within 1 kb for Enhancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Fetal DHS-1k</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.343788</td>\n",
       "      <td>0.228424</td>\n",
       "      <td>SNP pairs within 1 kb for Fetal DHS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>H3K27ac (Hnisz)-1k</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.337385</td>\n",
       "      <td>0.241750</td>\n",
       "      <td>SNP pairs within 1 kb for H3K27ac (Hnisz)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>H3K27ac (PGC2)-1k</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.327117</td>\n",
       "      <td>0.235396</td>\n",
       "      <td>SNP pairs within 1 kb for H3K27ac (PGC2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>H3K4me1-1k</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.342022</td>\n",
       "      <td>0.233561</td>\n",
       "      <td>SNP pairs within 1 kb for H3K4me1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>H3K4me3-1k</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.326912</td>\n",
       "      <td>0.215755</td>\n",
       "      <td>SNP pairs within 1 kb for H3K4me3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>H3K9ac-1k</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.358715</td>\n",
       "      <td>0.216737</td>\n",
       "      <td>SNP pairs within 1 kb for H3K9ac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Intron-1k</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.361928</td>\n",
       "      <td>0.231346</td>\n",
       "      <td>SNP pairs within 1 kb for Intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Promoter-1k</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.321337</td>\n",
       "      <td>0.206426</td>\n",
       "      <td>SNP pairs within 1 kb for Promoter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Repressed-1k</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.374997</td>\n",
       "      <td>0.268214</td>\n",
       "      <td>SNP pairs within 1 kb for Repressed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Super enhancer (Hnisz)-1k</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.306412</td>\n",
       "      <td>0.245634</td>\n",
       "      <td>SNP pairs within 1 kb for Super enhancer (Hnisz)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>TFBS-1k</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.306054</td>\n",
       "      <td>0.232053</td>\n",
       "      <td>SNP pairs within 1 kb for TFBS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Transcribed-1k</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.374280</td>\n",
       "      <td>0.222523</td>\n",
       "      <td>SNP pairs within 1 kb for Transcribed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Typical enhancer-1k</td>\n",
       "      <td>fct-1k</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.255268</td>\n",
       "      <td>0.269883</td>\n",
       "      <td>SNP pairs within 1 kb for Typical enhancer</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Name       Type  n_pair_common_common  n_pair_lf_lf  \\\n",
       "0         Proximal (0-0.1 kb)  proximity               2621179        981186   \n",
       "1           Proximal (0-1 kb)  proximity              23622561       8458072   \n",
       "2         Proximal (0.1-1 kb)  proximity              21001382       7476886   \n",
       "3                        Exon       gene                542754        262898   \n",
       "4                        Gene       gene            1367493711     521528493   \n",
       "5                 Exonic-gene       gene               1173399        608234   \n",
       "6              Protein-domain       gene                 93951         91867   \n",
       "7                    Promoter       gene                877113        365052   \n",
       "8         H3K27ac (Hnisz)-100    fct-100                     0             0   \n",
       "9          H3K27ac (PGC2)-100    fct-100                     0             0   \n",
       "10                H3K4me1-100    fct-100                     0             0   \n",
       "11                 Intron-100    fct-100                     0             0   \n",
       "12              Repressed-100    fct-100                     0             0   \n",
       "13            Transcribed-100    fct-100                     0             0   \n",
       "14                     DGF-1k     fct-1k                     0             0   \n",
       "15                     DHS-1k     fct-1k                     0             0   \n",
       "16                Enhancer-1k     fct-1k                     0             0   \n",
       "17               Fetal DHS-1k     fct-1k                     0             0   \n",
       "18         H3K27ac (Hnisz)-1k     fct-1k                     0             0   \n",
       "19          H3K27ac (PGC2)-1k     fct-1k                     0             0   \n",
       "20                 H3K4me1-1k     fct-1k                     0             0   \n",
       "21                 H3K4me3-1k     fct-1k                     0             0   \n",
       "22                  H3K9ac-1k     fct-1k                     0             0   \n",
       "23                  Intron-1k     fct-1k                     0             0   \n",
       "24                Promoter-1k     fct-1k                     0             0   \n",
       "25               Repressed-1k     fct-1k                     0             0   \n",
       "26  Super enhancer (Hnisz)-1k     fct-1k                     0             0   \n",
       "27                    TFBS-1k     fct-1k                     0             0   \n",
       "28             Transcribed-1k     fct-1k                     0             0   \n",
       "29        Typical enhancer-1k     fct-1k                     0             0   \n",
       "\n",
       "    avgr_common_common  avgr_lf_lf  \\\n",
       "0             0.416440    0.309336   \n",
       "1             0.364336    0.248938   \n",
       "2             0.357821    0.240968   \n",
       "3             0.278384    0.137887   \n",
       "4             0.065348    0.046902   \n",
       "5             0.204768    0.097622   \n",
       "6             0.205562    0.072658   \n",
       "7             0.217566    0.132904   \n",
       "8             0.396433    0.304018   \n",
       "9             0.386332    0.301221   \n",
       "10            0.400148    0.293532   \n",
       "11            0.416202    0.295014   \n",
       "12            0.425160    0.325253   \n",
       "13            0.424660    0.296087   \n",
       "14            0.348136    0.243331   \n",
       "15            0.355067    0.235842   \n",
       "16            0.350655    0.230790   \n",
       "17            0.343788    0.228424   \n",
       "18            0.337385    0.241750   \n",
       "19            0.327117    0.235396   \n",
       "20            0.342022    0.233561   \n",
       "21            0.326912    0.215755   \n",
       "22            0.358715    0.216737   \n",
       "23            0.361928    0.231346   \n",
       "24            0.321337    0.206426   \n",
       "25            0.374997    0.268214   \n",
       "26            0.306412    0.245634   \n",
       "27            0.306054    0.232053   \n",
       "28            0.374280    0.222523   \n",
       "29            0.255268    0.269883   \n",
       "\n",
       "                                         Description  \n",
       "0                                      Within 100 bp  \n",
       "1                                        Within 1 kb  \n",
       "2                            Between 100 bp and 1 kb  \n",
       "3                                          Same exon  \n",
       "4                                          Same gene  \n",
       "5                     Exonic region of the same gene  \n",
       "6                  Same protein domain and same gene  \n",
       "7     Pairs of promoter SNPs linked to the same gene  \n",
       "8        SNP pairs within 100 bp for H3K27ac (Hnisz)  \n",
       "9         SNP pairs within 100 bp for H3K27ac (PGC2)  \n",
       "10               SNP pairs within 100 bp for H3K4me1  \n",
       "11                SNP pairs within 100 bp for Intron  \n",
       "12             SNP pairs within 100 bp for Repressed  \n",
       "13           SNP pairs within 100 bp for Transcribed  \n",
       "14                     SNP pairs within 1 kb for DGF  \n",
       "15                     SNP pairs within 1 kb for DHS  \n",
       "16                SNP pairs within 1 kb for Enhancer  \n",
       "17               SNP pairs within 1 kb for Fetal DHS  \n",
       "18         SNP pairs within 1 kb for H3K27ac (Hnisz)  \n",
       "19          SNP pairs within 1 kb for H3K27ac (PGC2)  \n",
       "20                 SNP pairs within 1 kb for H3K4me1  \n",
       "21                 SNP pairs within 1 kb for H3K4me3  \n",
       "22                  SNP pairs within 1 kb for H3K9ac  \n",
       "23                  SNP pairs within 1 kb for Intron  \n",
       "24                SNP pairs within 1 kb for Promoter  \n",
       "25               SNP pairs within 1 kb for Repressed  \n",
       "26  SNP pairs within 1 kb for Super enhancer (Hnisz)  \n",
       "27                    SNP pairs within 1 kb for TFBS  \n",
       "28             SNP pairs within 1 kb for Transcribed  \n",
       "29        SNP pairs within 1 kb for Typical enhancer  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_PANNOT[['Name', 'Type', 'n_pair_common_common', 'n_pair_lf_lf', 'avgr_common_common', 'avgr_lf_lf', 'Description']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Old code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# DIC_NAME_pAN = {\n",
    "#     'proxy_0_100' : ['Proximal (0-0.1 kb)', 'All pairs of SNPs within 100 bp'],\n",
    "#     'proxy_0_1000' : ['Proximal (0-1 kb)', 'All pairs of SNPs within 1 kb'],\n",
    "#     'proxy_0_10000' : ['Proximal (0-10 kb)', 'All pairs of SNPs within 10 kb'],\n",
    "#     'proxy_100_1000' : ['Proximal (0.1-1 kb)', 'All pairs of SNPs between 100 bp and 1 kb'],\n",
    "#     'proxy_1000_10000' : ['Proximal (1-10 kb)', 'All pairs of SNPs between 1 kb and 10 kb'],\n",
    "#     'ldp5_proxy_10000' : ['High-LD', 'All pairs of SNPs with LD >= 0.5 and dist < 10 kb'],\n",
    "#     'ldp5_proxy_0_100' : ['High-LD (0-0.1 kb)', 'All pairs of SNPs with LD >= 0.5 and dist < 100 bp'],\n",
    "#     'ldp5_proxy_100_1000' : ['High-LD (0.1-1 kb)', \n",
    "#                              'All pairs of SNPs with LD >= 0.5 and dist between 100 bp and 1 kb'],\n",
    "#     'ldp5_proxy_1000_10000' : ['High-LD (1-10 kb)', \n",
    "#                                       'All pairs of SNPs with LD >= 0.5 and dist between 1 kb and 10 kb'],\n",
    "#     'exon' : ['Exon', 'All pairs of SNPs in the same exon'],\n",
    "#     'gene' : ['Gene', 'All pairs of SNPs in the same gene'],\n",
    "#     'exonic_gene' : ['Exonic-gene', 'All pair of exonic SNPs in the same gene body'],\n",
    "#     'protein_domain' : ['Protein-domain', 'All pair of SNPs in the same protein domain and same gene'],\n",
    "#     'cS2G_all' : ['cS2G-all', 'All pairs of SNPs linked to the same gene by cS2G'],\n",
    "#     'cS2G_promoter' : ['Promoter', 'All pairs of promoter SNPs linked to the same gene'],\n",
    "#     'cS2G_other' : [\n",
    "#         'cS2G-other', \n",
    "#         'All pairs of other SNPs linked to the same gene by cS2G, excluding exonic or promoter SNPs'],\n",
    "# }"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
