{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import time\n",
    "import os\n",
    "import gdreg\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pickle\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "\n",
    "# autoreload\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Trait info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DIC_NAME = {\n",
    "    'blood_PLATELET_COUNT': 'Platelet Count', \n",
    "    'blood_RBC_DISTRIB_WIDTH': 'Red Blood Cell Distribution Width',\n",
    "    'blood_RED_COUNT': 'Red Blood Cell Count',\n",
    "    'blood_WHITE_COUNT': 'White Blood Cell Count', \n",
    "    'bmd_HEEL_TSCOREz': 'Heel T Score', \n",
    "    'body_BALDING1': 'Balding Type I', \n",
    "    'body_BMIz': 'BMI',\n",
    "    'body_HEIGHTz': 'Height',\n",
    "    'body_WHRadjBMIz': 'Waist-hip Ratio',\n",
    "    'bp_DIASTOLICadjMEDz': 'Diastolic Blood Pressure',\n",
    "    'cov_EDU_YEARS': 'Years of Education',\n",
    "    'disease_ALLERGY_ECZEMA_DIAGNOSED': 'Eczema',\n",
    "    'lung_FEV1FVCzSMOKE': 'FEV1-FVC Ratio',\n",
    "    'lung_FVCzSMOKE': 'Forced Vital Capacity (FVC)',\n",
    "    'mental_NEUROTICISM': 'Neuroticism',\n",
    "    'other_MORNINGPERSON': 'Morning Person',\n",
    "    'pigment_SUNBURN': 'Sunburn Occasion',\n",
    "    'repro_MENARCHE_AGE': 'Age at Menarche',\n",
    "    'repro_MENOPAUSE_AGE': 'Age at Menopause',\n",
    "    'repro_NumberChildrenEverBorn_Pooled': 'Number Children (Pooled)',\n",
    "    'cancer_BREAST' : 'Breast Cancer',\n",
    "    'cancer_PROSTATE' : 'Prostate Cancer',\n",
    "    'disease_HYPOTHYROIDISM_SELF_REP' : 'Hypothyroidism',\n",
    "    'biochemistry_AlkalinePhosphatase' : 'Alkaline Phosphatase',\n",
    "    'biochemistry_AspartateAminotransferase' : 'Aspartate Aminotransferase',\n",
    "    'biochemistry_Cholesterol' : 'Cholesterol', \n",
    "    'biochemistry_Creatinine' : 'Creatinine',\n",
    "    'biochemistry_IGF1' : 'IGF1',\n",
    "    'biochemistry_Phosphate' : 'Phosphate',\n",
    "    'biochemistry_TotalBilirubin' : 'Total Bilirubin',\n",
    "    'biochemistry_TotalProtein' : 'Total Protein',\n",
    "    'biochemistry_VitaminD' : 'Vitamin D',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_info = pd.DataFrame(index=list(DIC_NAME))\n",
    "df_info['Trait Name'] = [DIC_NAME[x] for x in df_info.index]\n",
    "df_info['Trait_Identifier'] = df_info.index\n",
    "\n",
    "sumstats_file = '/n/groups/price/martin/data_GDREG/UKBimp_337K_MAF001/UKBB_trait/@.sumstats.gz'\n",
    "df_info['N.ukbimp'] = [pd.read_csv(sumstats_file.replace('@', x), nrows=10, sep='\\t')['N'].median()\n",
    "                for x in df_info.index]\n",
    "df_info['N.ukbimp'] = df_info['N.ukbimp'].astype(int)\n",
    "\n",
    "df_info.to_csv('/n/groups/price/martin/data_GDREG/results/tables/trait_info.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Single-SNP annotation info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# [Name, type, reference]\n",
    "DIC_NAME_AN = {\n",
    "    'BivFlnk' : ['Flanking Bivalent TSS/Enhancer', 'b', 'Roadmap Consortium 2015'],\n",
    "    'Conserved_LindbladToh' : ['Conserved (Lindblad-Toh)', 'b', 'Ward and Kellis 2012'],\n",
    "    'Conserved_Vertebrate_phastCons46way' : ['Conserved (phastCons - Vertebrate)', 'b', 'Siepel et al. 2005'],\n",
    "    'Conserved_Mammal_phastCons46way' : ['Conserved (phastCons - Mammal)', 'b', 'Siepel et al. 2005'],\n",
    "    'Conserved_Primate_phastCons46way' : ['Conserved (phastCons - Primate)', 'b', 'Siepel et al. 2005'],\n",
    "    'Coding_UCSC' : ['Coding', 'b', 'UCSC'],\n",
    "    'UTR_3_UCSC' : ['3UTR', 'b', 'UCSC'],\n",
    "    'UTR_5_UCSC' : ['5UTR', 'b', 'UCSC'],\n",
    "    'Intron_UCSC' : ['Intron', 'b', 'UCSC'],\n",
    "    'Enhancer_Andersson' : ['FANTOM5 Enhancer', 'b', 'Andersson et al. 2014'],\n",
    "    'SuperEnhancer_Hnisz' : ['Super Enhancer (Hnisz)', 'b', 'Hnisz et al. 2013'],\n",
    "    'Enhancer_Hoffman' : ['Enhancer', 'b', 'Hoffman et al. 2013'],\n",
    "    'WeakEnhancer_Hoffman' : ['Weak Enhancer', 'b', 'Hoffman et al. 2013'],\n",
    "    'Promoter_UCSC' : ['Promoter', 'b', 'UCSC'],\n",
    "    'PromoterFlanking_Hoffman' : ['Promoter Flanking', 'b', 'Hoffman et al. 2013'],\n",
    "    'DHS_Trynka' : ['DHS', 'b', 'Trynka et al. 2013'],\n",
    "    'FetalDHS_Trynka' : ['Fetal DHS', 'b', 'Trynka et al. 2013'],\n",
    "    'H3K27ac_Hnisz' : ['H3K27ac (Hnisz)', 'b', 'Hnisz et al. 2013'],\n",
    "    'H3K27ac_PGC2' : ['H3K27ac (PGC2)', 'b', 'PGC 2014'],\n",
    "    'H3K4me1_Trynka' : ['H3K4me1', 'b', 'Trynka et al. 2013'],\n",
    "    'H3K4me3_Trynka' : ['H3K4me3', 'b', 'Trynka et al. 2013'],\n",
    "    'H3K9ac_Trynka' : ['H3K9ac', 'b', 'Trynka et al. 2013'],\n",
    "    'TFBS_ENCODE' : ['TFBS', 'b', 'ENCODE'],\n",
    "    'Transcribed_Hoffman' : ['Transcribed', 'b', 'Hoffman et al. 2013'],\n",
    "    'TSS_Hoffman' : ['TSS', 'b', 'Hoffman et al. 2013'],\n",
    "    'CTCF_Hoffman' : ['CTCF', 'b', 'Hoffman et al. 2013'],\n",
    "    'DGF_ENCODE' : ['DGF', 'b', 'ENCODE'],\n",
    "    'Repressed_Hoffman' : ['Repressed', 'b', 'Hoffman et al. 2013'],\n",
    "    'Vahedi_Tcell_SE' : ['Super Enhancer (Vahedi)', 'b', 'Vahedi et al. 2015'],\n",
    "    'Vahedi_Tcell_TE' : ['Typical Enhancer', 'b', 'Vahedi et al. 2015'],\n",
    "    'GerpRS_g4' : ['Conserved (GERP RS >= 4)', 'b', 'Davydov et al. 2010'], # from CADD\n",
    "    'GerpN' : ['Conserved (GERP NS)', 'c', 'Davydov et al. 2010'], # from CADD\n",
    "    'alleleage' : ['MAF-adjusted predicted allele age', 'c', 'Rasmussen et al. 2014'],\n",
    "    # MAF-adjusted LLD-AFR removed\n",
    "    'nucleotide_div' : ['Nucleotide diversity', 'c', ''],\n",
    "    'recomb_rate' : ['Recombination Rate', 'c', 'HapMap2 map'],\n",
    "    'Backgrd_Selection_Stat' : ['McVicker B statistics', 'c', 'McVicker et al. 2009'],\n",
    "    'CpG' : ['CpG-Content', 'c', ''], # from CADD\n",
    "    # Non-synonymous removed\n",
    "    # Synonymous removed\n",
    "    'CADD_g20' : ['Deleterious (CADD >= 20)', 'b', 'Kircher et al. 2014'],\n",
    "#     'snpeff_downstream_gene_variant' : ['Downstream Gene', 'b', 'Cingolani et al. 2012'],\n",
    "#     'snpeff_non_coding_transcript_exon_variant' : ['Non-coding Exon', 'b', 'Cingolani et al. 2012'],\n",
    "    'snpeff_nonsynonymous_variant' : ['Non-synonymous', 'b', 'Cingolani et al. 2012'],\n",
    "#     'snpeff_splice_region_variant' : ['Splice Region', 'b', 'Cingolani et al. 2012'],\n",
    "    'snpeff_synonymous_variant' : ['Synonymous', 'b', 'Cingolani et al. 2012'],\n",
    "#     'snpeff_upstream_gene_variant' : ['Upstream Gene', 'b', 'Cingolani et al. 2012'],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_type(line):\n",
    "    if line == 'b':\n",
    "        return 'Binary'\n",
    "    if line == 'c':\n",
    "        return 'Continuous'\n",
    "\n",
    "df_info = pd.DataFrame(index=DIC_NAME_AN)\n",
    "df_info['Name'] = [DIC_NAME_AN[x][0] for x in df_info.index]\n",
    "df_info['Identifier'] = df_info.index\n",
    "df_info['Type'] = [get_type(DIC_NAME_AN[x][1]) for x in df_info.index]\n",
    "df_info['Reference/dataset'] = [DIC_NAME_AN[x][2] for x in df_info.index]\n",
    "\n",
    "# Version of the baseline model\n",
    "df_ldsc = pd.read_excel('/n/groups/price/martin/data_GDREG/results/Gazal_NG_2018_table.xlsx',\n",
    "                        sheet_name='S1', skiprows=1)\n",
    "df_ldsc.index = df_ldsc['Main annotation']\n",
    "df_ldsc.loc[df_ldsc['Version of the baseline model']=='New annotation', 'Version of the baseline model'] = \\\n",
    "    'Gazal et al. 2018' \n",
    "df_info['Version of the baseline model'] = [\n",
    "    df_ldsc.loc[x, 'Version of the baseline model'] \n",
    "    if (x in df_ldsc.index) & (x not in ['Non-synonymous', 'Synonymous']) \n",
    "    else 'New annotation' \n",
    "    for x in df_info['Name']\n",
    "]\n",
    "\n",
    "df_info.to_csv('/n/groups/price/martin/data_GDREG/results/tables/annot_info.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SNP-pair annotation info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "DIC_NAME_pAN = {\n",
    "    'proxy_0_100' : ['Proximal (0-0.1 kb)', 'All pairs of SNPs within 100 bp'],\n",
    "    'proxy_100_1000' : ['Proximal (0.1-1 kb)', 'All pairs of SNPs between 100 bp and 1 kb'],\n",
    "    'proxy_1000_10000' : ['Proximal (1-10 kb)', 'All pairs of SNPs between 1 kb and 10 kb'],\n",
    "    'ldp5_proxy_10000' : ['High-LD', 'All pairs of SNPs with LD >= 0.5 and dist < 10 kb'],\n",
    "    'pannot_ldp5_proxy_0_100' : ['High-LD (0-0.1 kb)', 'All pairs of SNPs with LD >= 0.5 and dist < 100 bp'],\n",
    "    'pannot_ldp5_proxy_100_1000' : ['High-LD (0.1-1 kb)', \n",
    "                                    'All pairs of SNPs with LD >= 0.5 and dist between 100 bp and 1 kb'],\n",
    "    'pannot_ldp5_proxy_1000_10000' : ['High-LD (1-10 kb)', \n",
    "                                      'All pairs of SNPs with LD >= 0.5 and dist between 1 kb and 10 kb'],\n",
    "    'exon' : ['Exon', 'All pairs of SNPs in the same exon'],\n",
    "    'gene' : ['Gene', 'All pairs of SNPs in the same gene'],\n",
    "    'exonic_gene' : ['Exonic-gene', 'All pair of exonic SNPs in the same gene body'],\n",
    "    'protein_domain' : ['Protein-domain', 'All pair of SNPs in the same protein domain and same gene'],\n",
    "    'cS2G_all' : ['cS2G-all', 'All pairs of SNPs linked to the same gene by cS2G'],\n",
    "    'cS2G_promoter' : ['cS2G-promoter', 'All pairs of promoter SNPs linked to the same gene by cS2G'],\n",
    "    'cS2G_other' : [\n",
    "        'cS2G-other', \n",
    "        'All pairs of other SNPs linked to the same gene by cS2G,  excluding exonic or promoter SNPs'],\n",
    "    'Dixon_2015_H1_ESC' : ['Loop-ESC', 'Chromatin loops in H1 human embryonic stem (ES) cells'],\n",
    "    'Dixon_2015_H1_MES' : ['Loop-MES', 'Chromatin loops in H1 human mesendoderm (ME) cell'],\n",
    "    'Dixon_2015_H1_MSC' : ['Loop-MSC', 'Chromatin loops in H1 human mesenchymal stem (MS) cell'],\n",
    "    'Dixon_2015_H1_NPC' : ['Loop-NPC', 'Chromatin loops in H1 human neural progenitor (NP) cell'],\n",
    "    'Dixon_2015_H1_TRO' : ['Loop-TRO', 'Chromatin loops in H1 human trophoblast-like (TB) cell'],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_info = pd.DataFrame(index=DIC_NAME_pAN)\n",
    "df_info['Name'] = [DIC_NAME_pAN[x][0] for x in df_info.index]\n",
    "df_info['Identifier'] = df_info.index\n",
    "df_info['Description'] = [DIC_NAME_pAN[x][1] for x in df_info.index]\n",
    "\n",
    "df_info.to_csv('/n/groups/price/martin/data_GDREG/results/tables/pannot_info.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Identifier</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>proxy_0_100</th>\n",
       "      <td>Proximal (0-0.1 kb)</td>\n",
       "      <td>proxy_0_100</td>\n",
       "      <td>All pairs of SNPs within 100 bp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>proxy_100_1000</th>\n",
       "      <td>Proximal (0.1-1 kb)</td>\n",
       "      <td>proxy_100_1000</td>\n",
       "      <td>All pairs of SNPs between 100 bp and 1 kb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>proxy_1000_10000</th>\n",
       "      <td>Proximal (1-10 kb)</td>\n",
       "      <td>proxy_1000_10000</td>\n",
       "      <td>All pairs of SNPs between 1 kb and 10 kb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ldp5_proxy_10000</th>\n",
       "      <td>High-LD</td>\n",
       "      <td>ldp5_proxy_10000</td>\n",
       "      <td>All pairs of SNPs with LD &gt;= 0.5 and dist &lt; 10 kb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pannot_ldp5_proxy_0_100</th>\n",
       "      <td>High-LD (0-0.1 kb)</td>\n",
       "      <td>pannot_ldp5_proxy_0_100</td>\n",
       "      <td>All pairs of SNPs with LD &gt;= 0.5 and dist &lt; 10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pannot_ldp5_proxy_100_1000</th>\n",
       "      <td>High-LD (0.1-1 kb)</td>\n",
       "      <td>pannot_ldp5_proxy_100_1000</td>\n",
       "      <td>All pairs of SNPs with LD &gt;= 0.5 and dist betw...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pannot_ldp5_proxy_1000_10000</th>\n",
       "      <td>High-LD (1-10 kb)</td>\n",
       "      <td>pannot_ldp5_proxy_1000_10000</td>\n",
       "      <td>All pairs of SNPs with LD &gt;= 0.5 and dist betw...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon</th>\n",
       "      <td>Exon</td>\n",
       "      <td>exon</td>\n",
       "      <td>All pairs of SNPs in the same exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <td>Gene</td>\n",
       "      <td>gene</td>\n",
       "      <td>All pairs of SNPs in the same gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exonic_gene</th>\n",
       "      <td>Exonic-gene</td>\n",
       "      <td>exonic_gene</td>\n",
       "      <td>All pair of exonic SNPs in the same gene body</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_domain</th>\n",
       "      <td>Protein-domain</td>\n",
       "      <td>protein_domain</td>\n",
       "      <td>All pair of SNPs in the same protein domain an...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cS2G_all</th>\n",
       "      <td>cS2G-all</td>\n",
       "      <td>cS2G_all</td>\n",
       "      <td>All pairs of SNPs linked to the same gene by cS2G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cS2G_promoter</th>\n",
       "      <td>cS2G-promoter</td>\n",
       "      <td>cS2G_promoter</td>\n",
       "      <td>All pairs of promoter SNPs linked to the same ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cS2G_other</th>\n",
       "      <td>cS2G-other</td>\n",
       "      <td>cS2G_other</td>\n",
       "      <td>All pairs of other SNPs linked to the same gen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dixon_2015_H1_ESC</th>\n",
       "      <td>Loop-ESC</td>\n",
       "      <td>Dixon_2015_H1_ESC</td>\n",
       "      <td>Chromatin loops in H1 human embryonic stem (ES...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dixon_2015_H1_MES</th>\n",
       "      <td>Loop-MES</td>\n",
       "      <td>Dixon_2015_H1_MES</td>\n",
       "      <td>Chromatin loops in H1 human mesendoderm (ME) cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dixon_2015_H1_MSC</th>\n",
       "      <td>Loop-MSC</td>\n",
       "      <td>Dixon_2015_H1_MSC</td>\n",
       "      <td>Chromatin loops in H1 human mesenchymal stem (...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dixon_2015_H1_NPC</th>\n",
       "      <td>Loop-NPC</td>\n",
       "      <td>Dixon_2015_H1_NPC</td>\n",
       "      <td>Chromatin loops in H1 human neural progenitor ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dixon_2015_H1_TRO</th>\n",
       "      <td>Loop-TRO</td>\n",
       "      <td>Dixon_2015_H1_TRO</td>\n",
       "      <td>Chromatin loops in H1 human trophoblast-like (...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Name  \\\n",
       "proxy_0_100                   Proximal (0-0.1 kb)   \n",
       "proxy_100_1000                Proximal (0.1-1 kb)   \n",
       "proxy_1000_10000               Proximal (1-10 kb)   \n",
       "ldp5_proxy_10000                          High-LD   \n",
       "pannot_ldp5_proxy_0_100        High-LD (0-0.1 kb)   \n",
       "pannot_ldp5_proxy_100_1000     High-LD (0.1-1 kb)   \n",
       "pannot_ldp5_proxy_1000_10000    High-LD (1-10 kb)   \n",
       "exon                                         Exon   \n",
       "gene                                         Gene   \n",
       "exonic_gene                           Exonic-gene   \n",
       "protein_domain                     Protein-domain   \n",
       "cS2G_all                                 cS2G-all   \n",
       "cS2G_promoter                       cS2G-promoter   \n",
       "cS2G_other                             cS2G-other   \n",
       "Dixon_2015_H1_ESC                        Loop-ESC   \n",
       "Dixon_2015_H1_MES                        Loop-MES   \n",
       "Dixon_2015_H1_MSC                        Loop-MSC   \n",
       "Dixon_2015_H1_NPC                        Loop-NPC   \n",
       "Dixon_2015_H1_TRO                        Loop-TRO   \n",
       "\n",
       "                                                Identifier  \\\n",
       "proxy_0_100                                    proxy_0_100   \n",
       "proxy_100_1000                              proxy_100_1000   \n",
       "proxy_1000_10000                          proxy_1000_10000   \n",
       "ldp5_proxy_10000                          ldp5_proxy_10000   \n",
       "pannot_ldp5_proxy_0_100            pannot_ldp5_proxy_0_100   \n",
       "pannot_ldp5_proxy_100_1000      pannot_ldp5_proxy_100_1000   \n",
       "pannot_ldp5_proxy_1000_10000  pannot_ldp5_proxy_1000_10000   \n",
       "exon                                                  exon   \n",
       "gene                                                  gene   \n",
       "exonic_gene                                    exonic_gene   \n",
       "protein_domain                              protein_domain   \n",
       "cS2G_all                                          cS2G_all   \n",
       "cS2G_promoter                                cS2G_promoter   \n",
       "cS2G_other                                      cS2G_other   \n",
       "Dixon_2015_H1_ESC                        Dixon_2015_H1_ESC   \n",
       "Dixon_2015_H1_MES                        Dixon_2015_H1_MES   \n",
       "Dixon_2015_H1_MSC                        Dixon_2015_H1_MSC   \n",
       "Dixon_2015_H1_NPC                        Dixon_2015_H1_NPC   \n",
       "Dixon_2015_H1_TRO                        Dixon_2015_H1_TRO   \n",
       "\n",
       "                                                                    Description  \n",
       "proxy_0_100                                     All pairs of SNPs within 100 bp  \n",
       "proxy_100_1000                        All pairs of SNPs between 100 bp and 1 kb  \n",
       "proxy_1000_10000                       All pairs of SNPs between 1 kb and 10 kb  \n",
       "ldp5_proxy_10000              All pairs of SNPs with LD >= 0.5 and dist < 10 kb  \n",
       "pannot_ldp5_proxy_0_100       All pairs of SNPs with LD >= 0.5 and dist < 10...  \n",
       "pannot_ldp5_proxy_100_1000    All pairs of SNPs with LD >= 0.5 and dist betw...  \n",
       "pannot_ldp5_proxy_1000_10000  All pairs of SNPs with LD >= 0.5 and dist betw...  \n",
       "exon                                         All pairs of SNPs in the same exon  \n",
       "gene                                         All pairs of SNPs in the same gene  \n",
       "exonic_gene                       All pair of exonic SNPs in the same gene body  \n",
       "protein_domain                All pair of SNPs in the same protein domain an...  \n",
       "cS2G_all                      All pairs of SNPs linked to the same gene by cS2G  \n",
       "cS2G_promoter                 All pairs of promoter SNPs linked to the same ...  \n",
       "cS2G_other                    All pairs of other SNPs linked to the same gen...  \n",
       "Dixon_2015_H1_ESC             Chromatin loops in H1 human embryonic stem (ES...  \n",
       "Dixon_2015_H1_MES             Chromatin loops in H1 human mesendoderm (ME) cell  \n",
       "Dixon_2015_H1_MSC             Chromatin loops in H1 human mesenchymal stem (...  \n",
       "Dixon_2015_H1_NPC             Chromatin loops in H1 human neural progenitor ...  \n",
       "Dixon_2015_H1_TRO             Chromatin loops in H1 human trophoblast-like (...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
