{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gencode exon & gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_gencode = pd.read_csv(\n",
    "#     \"/n/groups/price/martin/data_GDREG/gene_annotation/gencode.v41lift37.basic.annotation.gtf.gz\", \n",
    "#     skiprows=5, sep='\\t', header=None\n",
    "# )\n",
    "\n",
    "df_gene_cs2g = pd.read_csv(\n",
    "    \"/n/groups/price/martin/data_GDREG/gene_annotation/cS2G/list_genes_qc.txt.gz\", sep='\\t'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Get gene annotations 63678\n",
      "Retain annots in CHRs 1-22 60472\n",
      "Drop duplicates 60472\n",
      "Restrict to Gazel et al. 19,995 genes:  19606\n"
     ]
    }
   ],
   "source": [
    "def get_gene_info(line, term='gene_id'):\n",
    "    \"\"\" term is one of ['gene_id', 'gene_type', 'gene_name']\n",
    "    \"\"\"    \n",
    "    for info in line.split(';'):\n",
    "        info = info.strip()\n",
    "        if info.startswith(term):\n",
    "            return info.replace(term, '').replace('\\\"', '').strip()\n",
    "\n",
    "df_gene = df_gencode.loc[df_gencode[2]=='gene'].copy()\n",
    "df_gene['ENSG'] = [get_gene_info(x, term='gene_id').split('.')[0] for x in df_gene[8]]\n",
    "df_gene['GENE_TYPE'] = [get_gene_info(x, term='gene_type').split('.')[0] for x in df_gene[8]]\n",
    "df_gene['GENE_NAME'] = [get_gene_info(x, term='gene_name').split('.')[0] for x in df_gene[8]]\n",
    "print(\"Get gene annotations\", df_gene.shape[0])\n",
    "\n",
    "df_gene = df_gene.loc[~df_gene[0].isin(['chrX', 'chrY', 'chrM'])].copy()\n",
    "df_gene['CHR'] = df_gene[0]\n",
    "df_gene['START'] = df_gene[3]\n",
    "df_gene['END'] = df_gene[4] \n",
    "print(\"Retain annots in CHRs 1-22\", df_gene.shape[0])\n",
    "\n",
    "df_gene.drop_duplicates(subset=['CHR', 'START', 'END', 'ENSG', 'GENE_TYPE', 'GENE_NAME'], inplace=True)\n",
    "print(\"Drop duplicates\", df_gene.shape[0])\n",
    "\n",
    "gene_set = set(df_gene_cs2g['ENSG'])\n",
    "df_gene = df_gene.loc[df_gene['ENSG'].isin(gene_set)]\n",
    "print(\"Restrict to Gazel et al. 19,995 genes: \", df_gene.shape[0])\n",
    "\n",
    "df_gene = df_gene[['CHR', 'START', 'END', 'ENSG', 'GENE_TYPE', 'GENE_NAME']].copy()\n",
    "df_gene.to_csv('/n/groups/price/martin/data_GDREG/gene_annotation/ENSG_gene_annot_v41.txt',\n",
    "               sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Get exon annotations 840632\n",
      "Retain annots in CHRs 1-22 805376\n",
      "Drop duplicates 411792\n",
      "Restrict exons to Gazel et al. 19,995 genes:  297521\n"
     ]
    }
   ],
   "source": [
    "def get_exon_info(line, term='exon_id'):\n",
    "    \"\"\" term is one of ['exon_id', 'gene_id', 'gene_type', 'gene_name']\n",
    "    \"\"\"    \n",
    "    for info in line.split(';'):\n",
    "        info = info.strip()\n",
    "        if info.startswith(term):\n",
    "            return info.replace(term, '').replace('\\\"', '').strip()\n",
    "\n",
    "df_exon = df_gencode.loc[df_gencode[2]=='exon'].copy()\n",
    "df_exon['ENSE'] = [get_exon_info(x, term='exon_id').split('.')[0] for x in df_exon[8]]\n",
    "df_exon['ENSG'] = [get_exon_info(x, term='gene_id').split('.')[0] for x in df_exon[8]]\n",
    "df_exon['GENE_TYPE'] = [get_exon_info(x, term='gene_type') for x in df_exon[8]]\n",
    "df_exon['GENE_NAME'] = [get_exon_info(x, term='gene_name') for x in df_exon[8]]\n",
    "print(\"Get exon annotations\", df_exon.shape[0])\n",
    "\n",
    "df_exon = df_exon.loc[~df_exon[0].isin(['chrX', 'chrY', 'chrM'])].copy()\n",
    "df_exon['CHR'] = df_exon[0]\n",
    "df_exon['START'] = df_exon[3]\n",
    "df_exon['END'] = df_exon[4] \n",
    "print(\"Retain annots in CHRs 1-22\", df_exon.shape[0])\n",
    "\n",
    "df_exon.drop_duplicates(\n",
    "    subset=['CHR', 'START', 'END', 'ENSE', 'ENSG', 'GENE_TYPE', 'GENE_NAME'], inplace=True\n",
    ")\n",
    "print(\"Drop duplicates\", df_exon.shape[0])\n",
    "\n",
    "gene_set = set(df_gene_cs2g['ENSG'])\n",
    "df_exon = df_exon.loc[df_exon['ENSG'].isin(gene_set)]\n",
    "print(\"Restrict exons to Gazel et al. 19,995 genes: \", df_exon.shape[0])\n",
    "\n",
    "df_exon = df_exon[['CHR', 'START', 'END', 'ENSE', 'ENSG', 'GENE_TYPE', 'GENE_NAME']].copy()\n",
    "df_exon.to_csv('/n/groups/price/martin/data_GDREG/gene_annotation/ENSE_exon_annot_v41.txt',\n",
    "               sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### cS2G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "ANNOT_FILE = '/n/groups/price/martin/data_GDREG/gene_annotation/cS2G/cS2G_UKBB/cS2G.@.SGscore.gz'\n",
    "\n",
    "for CHR in [22]:\n",
    "    df_annot = pd.read_csv(ANNOT_FILE.replace('@', '%d' % CHR), sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>GENE</th>\n",
       "      <th>cS2G</th>\n",
       "      <th>INFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22:16258211_A_G</td>\n",
       "      <td>POTEH</td>\n",
       "      <td>1.0</td>\n",
       "      <td>|Exon=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22:16258290_CATT_C</td>\n",
       "      <td>POTEH</td>\n",
       "      <td>1.0</td>\n",
       "      <td>|Exon=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22:16258294_T_G</td>\n",
       "      <td>POTEH</td>\n",
       "      <td>1.0</td>\n",
       "      <td>|Exon=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22:16266920_C_CAG</td>\n",
       "      <td>POTEH</td>\n",
       "      <td>1.0</td>\n",
       "      <td>|Exon=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22:16279319_GCAAAAACGTATGTAATT_G</td>\n",
       "      <td>POTEH</td>\n",
       "      <td>1.0</td>\n",
       "      <td>|Exon=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113011</th>\n",
       "      <td>22:51239281_G_C</td>\n",
       "      <td>RPL23AP82</td>\n",
       "      <td>1.0</td>\n",
       "      <td>|Exon=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113012</th>\n",
       "      <td>22:51239296_T_C</td>\n",
       "      <td>RPL23AP82</td>\n",
       "      <td>1.0</td>\n",
       "      <td>|Exon=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113013</th>\n",
       "      <td>22:51239304_C_T</td>\n",
       "      <td>RPL23AP82</td>\n",
       "      <td>1.0</td>\n",
       "      <td>|Exon=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113014</th>\n",
       "      <td>22:51239586_T_G</td>\n",
       "      <td>RPL23AP82</td>\n",
       "      <td>1.0</td>\n",
       "      <td>|Exon=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113015</th>\n",
       "      <td>22:51240820_C_T</td>\n",
       "      <td>MAPK8IP2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>|GTeX_Finemapped=1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>113016 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     SNP       GENE  cS2G                INFO\n",
       "0                        22:16258211_A_G      POTEH   1.0             |Exon=1\n",
       "1                     22:16258290_CATT_C      POTEH   1.0             |Exon=1\n",
       "2                        22:16258294_T_G      POTEH   1.0             |Exon=1\n",
       "3                      22:16266920_C_CAG      POTEH   1.0             |Exon=1\n",
       "4       22:16279319_GCAAAAACGTATGTAATT_G      POTEH   1.0             |Exon=1\n",
       "...                                  ...        ...   ...                 ...\n",
       "113011                   22:51239281_G_C  RPL23AP82   1.0             |Exon=1\n",
       "113012                   22:51239296_T_C  RPL23AP82   1.0             |Exon=1\n",
       "113013                   22:51239304_C_T  RPL23AP82   1.0             |Exon=1\n",
       "113014                   22:51239586_T_G  RPL23AP82   1.0             |Exon=1\n",
       "113015                   22:51240820_C_T   MAPK8IP2   1.0  |GTeX_Finemapped=1\n",
       "\n",
       "[113016 rows x 4 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['|Exon=1', '|Promoter=1', '|GTeX_Finemapped=1', '|ABC=1',\n",
       "       '|EpiMap=1', '|GTeX_Finemapped=1|eQTLGen_Finemapped=1',\n",
       "       '|Promoter=1|ABC=1|Cicero=1',\n",
       "       '|Promoter=1|GTeX_Finemapped=1|ABC=1|Cicero=1',\n",
       "       '|Promoter=1|ABC=1', '|ABC=1|Cicero=1',\n",
       "       '|GTeX_Finemapped=1|ABC=1|Cicero=1', '|GTeX_Finemapped=1|ABC=1',\n",
       "       '|Exon=1|ABC=1', '|Exon=1|GTeX_Finemapped=1',\n",
       "       '|eQTLGen_Finemapped=1', '|Cicero=1',\n",
       "       '|eQTLGen_Finemapped=1|ABC=1', '|GTeX_Finemapped=1|EpiMap=1',\n",
       "       '|Exon=1|EpiMap=1', '|Exon=1|eQTLGen_Finemapped=1|EpiMap=1',\n",
       "       '|Exon=1|eQTLGen_Finemapped=1', '|EpiMap=1|ABC=1',\n",
       "       '|Promoter=1|eQTLGen_Finemapped=1', '|Promoter=1|Cicero=1',\n",
       "       '|Promoter=1|eQTLGen_Finemapped=1|Cicero=1', '|Exon=1|Cicero=1',\n",
       "       '|Promoter=1|GTeX_Finemapped=1|eQTLGen_Finemapped=1|ABC=1',\n",
       "       '|Promoter=1|GTeX_Finemapped=1|ABC=1',\n",
       "       '|GTeX_Finemapped=1|eQTLGen_Finemapped=1|ABC=1',\n",
       "       '|Exon=1|GTeX_Finemapped=1|eQTLGen_Finemapped=1',\n",
       "       '|Exon=1|GTeX_Finemapped=1|eQTLGen_Finemapped=1|ABC=1',\n",
       "       '|Promoter=1|EpiMap=1', '|Exon=1|GTeX_Finemapped=1|ABC=1',\n",
       "       '|Exon=1|EpiMap=1|ABC=1',\n",
       "       '|GTeX_Finemapped=1|eQTLGen_Finemapped=1|EpiMap=1',\n",
       "       '|Exon=1|GTeX_Finemapped=1|EpiMap=1',\n",
       "       '|Exon=1|GTeX_Finemapped=1|eQTLGen_Finemapped=1|EpiMap=1',\n",
       "       '|eQTLGen_Finemapped=1|EpiMap=1',\n",
       "       '|Exon=1|GTeX_Finemapped=1|EpiMap=1|ABC=1',\n",
       "       '|Promoter=1|GTeX_Finemapped=1|eQTLGen_Finemapped=1',\n",
       "       '|Promoter=1|GTeX_Finemapped=1',\n",
       "       '|Promoter=1|eQTLGen_Finemapped=1|ABC=1',\n",
       "       '|Promoter=1|EpiMap=1|ABC=1',\n",
       "       '|Promoter=1|eQTLGen_Finemapped=1|EpiMap=1',\n",
       "       '|Promoter=1|GTeX_Finemapped=1|EpiMap=1',\n",
       "       '|GTeX_Finemapped=1|EpiMap=1|ABC=1', '|Exon=1|ABC=1|Cicero=1',\n",
       "       '|EpiMap=0.500', '|Promoter=1|GTeX_Finemapped=1|EpiMap=1|ABC=1',\n",
       "       '|Promoter=0.500', '|Promoter=0.500|ABC=1',\n",
       "       '|Promoter=0.500|GTeX_Finemapped=1|ABC=1', '|EpiMap=0.500|ABC=1',\n",
       "       '|EpiMap=0.500|Cicero=1',\n",
       "       '|eQTLGen_Finemapped=1|EpiMap=1|Cicero=1',\n",
       "       '|EpiMap=1|ABC=1|Cicero=1', '|EpiMap=1|Cicero=1',\n",
       "       '|Exon=1|eQTLGen_Finemapped=1|ABC=1',\n",
       "       '|eQTLGen_Finemapped=1|EpiMap=1|ABC=1|Cicero=1',\n",
       "       '|eQTLGen_Finemapped=1|ABC=1|Cicero=1',\n",
       "       '|Promoter=1|GTeX_Finemapped=1|Cicero=1',\n",
       "       '|GTeX_Finemapped=1|Cicero=1',\n",
       "       '|Promoter=1|EpiMap=1|ABC=1|Cicero=1', '|EpiMap=1|ABC=0.500',\n",
       "       '|ABC=0.500', '|Promoter=1|ABC=0.500',\n",
       "       '|Promoter=1|GTeX_Finemapped=1|ABC=0.500', '|Exon=1|ABC=0.500',\n",
       "       '|Promoter=0.500|GTeX_Finemapped=1|ABC=0.500',\n",
       "       '|Promoter=0.500|ABC=0.500',\n",
       "       '|GTeX_Finemapped=1|eQTLGen_Finemapped=1|Cicero=1',\n",
       "       '|GTeX_Finemapped=1|ABC=0.500',\n",
       "       '|GTeX_Finemapped=1|eQTLGen_Finemapped=1|ABC=0.500',\n",
       "       '|eQTLGen_Finemapped=1|EpiMap=1|ABC=0.500',\n",
       "       '|Exon=1|EpiMap=1|ABC=0.500',\n",
       "       '|GTeX_Finemapped=1|EpiMap=1|ABC=0.500',\n",
       "       '|eQTLGen_Finemapped=1|ABC=0.500',\n",
       "       '|Promoter=0.500|GTeX_Finemapped=1', '|Promoter=0.500|Cicero=1',\n",
       "       '|Promoter=0.500|GTeX_Finemapped=1|Cicero=1',\n",
       "       '|GTeX_Finemapped=1|eQTLGen_Finemapped=1|EpiMap=1|ABC=1',\n",
       "       '|Promoter=1|GTeX_Finemapped=1|eQTLGen_Finemapped=1|EpiMap=1|ABC=1',\n",
       "       '|eQTLGen_Finemapped=1|EpiMap=1|ABC=1', '|Promoter=0.500|EpiMap=1',\n",
       "       '|Promoter=0.500|EpiMap=1|ABC=1',\n",
       "       '|Promoter=1|eQTLGen_Finemapped=1|ABC=1|Cicero=1',\n",
       "       '|Exon=1|eQTLGen_Finemapped=1|ABC=1|Cicero=1',\n",
       "       '|Exon=1|GTeX_Finemapped=1|ABC=1|Cicero=1',\n",
       "       '|Promoter=0.500|eQTLGen_Finemapped=1|ABC=1',\n",
       "       '|Exon=1|GTeX_Finemapped=1|EpiMap=1|Cicero=1',\n",
       "       '|Promoter=1|EpiMap=1|Cicero=1',\n",
       "       '|Exon=1|GTeX_Finemapped=1|eQTLGen_Finemapped=1|ABC=1|Cicero=1',\n",
       "       '|Exon=1|EpiMap=1|Cicero=1',\n",
       "       '|Exon=1|eQTLGen_Finemapped=1|ABC=0.500',\n",
       "       '|Promoter=1|eQTLGen_Finemapped=1|ABC=0.500',\n",
       "       '|Exon=1|GTeX_Finemapped=1|ABC=0.500',\n",
       "       '|eQTLGen_Finemapped=1|Cicero=1',\n",
       "       '|Promoter=0.500|eQTLGen_Finemapped=1',\n",
       "       '|Promoter=0.500|GTeX_Finemapped=1|eQTLGen_Finemapped=1',\n",
       "       '|Promoter=1|eQTLGen_Finemapped=1|EpiMap=1|ABC=1',\n",
       "       '|Promoter=1|GTeX_Finemapped=1|EpiMap=1|ABC=1|Cicero=1',\n",
       "       '|GTeX_Finemapped=1|EpiMap=1|Cicero=1',\n",
       "       '|Promoter=0.500|GTeX_Finemapped=1|EpiMap=1|ABC=1',\n",
       "       '|eQTLGen_Finemapped=0.500',\n",
       "       '|Promoter=1|GTeX_Finemapped=1|eQTLGen_Finemapped=1|EpiMap=1',\n",
       "       '|Exon=0.500',\n",
       "       '|GTeX_Finemapped=1|eQTLGen_Finemapped=1|EpiMap=1|ABC=0.500',\n",
       "       '|Promoter=1|GTeX_Finemapped=1|eQTLGen_Finemapped=1|ABC=0.500',\n",
       "       '|ABC=0.500|Cicero=1', '|Exon=1|ABC=0.500|Cicero=1',\n",
       "       '|Promoter=0.500|GTeX_Finemapped=1|eQTLGen_Finemapped=1|ABC=0.500',\n",
       "       '|GTeX_Finemapped=1|EpiMap=1|ABC=1|Cicero=1',\n",
       "       '|Promoter=0.500|GTeX_Finemapped=1|eQTLGen_Finemapped=1|ABC=1',\n",
       "       '|Exon=1|GTeX_Finemapped=1|eQTLGen_Finemapped=1|EpiMap=1|ABC=0.500',\n",
       "       '|Promoter=0.500|GTeX_Finemapped=1|eQTLGen_Finemapped=1|ABC=1|Cicero=1',\n",
       "       '|GTeX_Finemapped=0.500|EpiMap=1', '|GTeX_Finemapped=0.500|ABC=1',\n",
       "       '|Exon=1|EpiMap=1|ABC=1|Cicero=1',\n",
       "       '|GTeX_Finemapped=1|eQTLGen_Finemapped=0.500',\n",
       "       '|Exon=1|eQTLGen_Finemapped=1|EpiMap=1|ABC=1',\n",
       "       '|Exon=1|eQTLGen_Finemapped=1|EpiMap=1|ABC=1|Cicero=1',\n",
       "       '|GTeX_Finemapped=1|eQTLGen_Finemapped=1|EpiMap=1|ABC=1|Cicero=1',\n",
       "       '|Exon=1|GTeX_Finemapped=1|EpiMap=1|ABC=1|Cicero=1',\n",
       "       '|EpiMap=0.500|ABC=1|Cicero=1',\n",
       "       '|GTeX_Finemapped=1|eQTLGen_Finemapped=1|EpiMap=1|Cicero=1',\n",
       "       '|Promoter=1|GTeX_Finemapped=0.500|eQTLGen_Finemapped=1|ABC=1',\n",
       "       '|GTeX_Finemapped=0.500|Cicero=1', '|GTeX_Finemapped=0.333',\n",
       "       '|Exon=1|GTeX_Finemapped=1|eQTLGen_Finemapped=1|EpiMap=1|ABC=1|Cicero=1',\n",
       "       '|Promoter=0.500|ABC=1|Cicero=1',\n",
       "       '|Promoter=0.500|eQTLGen_Finemapped=1|ABC=0.500',\n",
       "       '|Exon=1|GTeX_Finemapped=1|Cicero=1',\n",
       "       '|Promoter=0.500|EpiMap=1|ABC=1|Cicero=1',\n",
       "       '|Promoter=0.500|GTeX_Finemapped=1|EpiMap=1',\n",
       "       '|Promoter=1|GTeX_Finemapped=1|eQTLGen_Finemapped=1|ABC=1|Cicero=1',\n",
       "       '|Exon=1|eQTLGen_Finemapped=1|Cicero=1',\n",
       "       '|Promoter=0.500|ABC=0.500|Cicero=1', '|Exon=1|EpiMap=0.500',\n",
       "       '|Promoter=1|EpiMap=1|ABC=0.500',\n",
       "       '|Promoter=0.500|GTeX_Finemapped=1|eQTLGen_Finemapped=1|EpiMap=1',\n",
       "       '|Exon=0.500|ABC=1',\n",
       "       '|Exon=1|GTeX_Finemapped=1|eQTLGen_Finemapped=1|Cicero=1',\n",
       "       '|Promoter=1|GTeX_Finemapped=1|eQTLGen_Finemapped=1|EpiMap=1|ABC=1|Cicero=1',\n",
       "       '|Promoter=1|GTeX_Finemapped=1|EpiMap=1|Cicero=1'], dtype=object)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_annot['INFO'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Chromatin loops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_loop_file(fpath):\n",
    "    temp_df = pd.read_csv(LOOP_FILE, sep='\\t', header=None)\n",
    "    temp_df.columns = ['CHR', 'START', 'END', 'CHR1', 'START1', 'END1']\n",
    "    chr_list = ['chr%d' % x for x in range(1,23)]\n",
    "    temp_df = temp_df.loc[temp_df['CHR'].isin(chr_list)].copy()\n",
    "    temp_df['CHR'] = [int(x.replace('chr','')) for x in temp_df['CHR']]\n",
    "    temp_df['CHR1'] = [int(x.replace('chr','')) for x in temp_df['CHR1']]\n",
    "    temp_df.sort_values(['CHR', 'START'], inplace=True)\n",
    "    return temp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "LOOP_FILE = '/n/groups/price/martin/data_GDREG/gene_annotation/3dgenome_loops/'\\\n",
    "    'Dixon_2015.H1-ESC.hg19.peakachu-merged.loops'\n",
    "temp_df0 = read_loop_file(LOOP_FILE)\n",
    "\n",
    "LOOP_FILE = '/n/groups/price/martin/data_GDREG/gene_annotation/3dgenome_loops/'\\\n",
    "    'Dixon_2015.H1-MES.hg19.peakachu-merged.loops'\n",
    "temp_df1 = read_loop_file(LOOP_FILE)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3782"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(temp_df0['START']) & set(temp_df1['START']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>START</th>\n",
       "      <th>END</th>\n",
       "      <th>CHR1</th>\n",
       "      <th>START1</th>\n",
       "      <th>END1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13928</th>\n",
       "      <td>1</td>\n",
       "      <td>4710060</td>\n",
       "      <td>4720060</td>\n",
       "      <td>1</td>\n",
       "      <td>4830060</td>\n",
       "      <td>4840060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3913</th>\n",
       "      <td>1</td>\n",
       "      <td>5570060</td>\n",
       "      <td>5580060</td>\n",
       "      <td>1</td>\n",
       "      <td>5660060</td>\n",
       "      <td>5670060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3945</th>\n",
       "      <td>1</td>\n",
       "      <td>5570060</td>\n",
       "      <td>5580060</td>\n",
       "      <td>1</td>\n",
       "      <td>6020060</td>\n",
       "      <td>6030060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4019</th>\n",
       "      <td>1</td>\n",
       "      <td>6960060</td>\n",
       "      <td>6970060</td>\n",
       "      <td>1</td>\n",
       "      <td>7270060</td>\n",
       "      <td>7280060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3951</th>\n",
       "      <td>1</td>\n",
       "      <td>7290060</td>\n",
       "      <td>7300060</td>\n",
       "      <td>1</td>\n",
       "      <td>7570060</td>\n",
       "      <td>7580060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12952</th>\n",
       "      <td>22</td>\n",
       "      <td>44035880</td>\n",
       "      <td>44045880</td>\n",
       "      <td>22</td>\n",
       "      <td>44135880</td>\n",
       "      <td>44145880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12946</th>\n",
       "      <td>22</td>\n",
       "      <td>45375880</td>\n",
       "      <td>45385880</td>\n",
       "      <td>22</td>\n",
       "      <td>45485881</td>\n",
       "      <td>45495881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12956</th>\n",
       "      <td>22</td>\n",
       "      <td>46045880</td>\n",
       "      <td>46055880</td>\n",
       "      <td>22</td>\n",
       "      <td>46235880</td>\n",
       "      <td>46245880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12962</th>\n",
       "      <td>22</td>\n",
       "      <td>49475812</td>\n",
       "      <td>49485812</td>\n",
       "      <td>22</td>\n",
       "      <td>49685926</td>\n",
       "      <td>49695925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12959</th>\n",
       "      <td>22</td>\n",
       "      <td>50183648</td>\n",
       "      <td>50193648</td>\n",
       "      <td>22</td>\n",
       "      <td>50538429</td>\n",
       "      <td>50548429</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15557 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CHR     START       END  CHR1    START1      END1\n",
       "13928    1   4710060   4720060     1   4830060   4840060\n",
       "3913     1   5570060   5580060     1   5660060   5670060\n",
       "3945     1   5570060   5580060     1   6020060   6030060\n",
       "4019     1   6960060   6970060     1   7270060   7280060\n",
       "3951     1   7290060   7300060     1   7570060   7580060\n",
       "...    ...       ...       ...   ...       ...       ...\n",
       "12952   22  44035880  44045880    22  44135880  44145880\n",
       "12946   22  45375880  45385880    22  45485881  45495881\n",
       "12956   22  46045880  46055880    22  46235880  46245880\n",
       "12962   22  49475812  49485812    22  49685926  49695925\n",
       "12959   22  50183648  50193648    22  50538429  50548429\n",
       "\n",
       "[15557 rows x 6 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp_df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ind_select.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
